<!DOCTYPE html>
<html>
<head>
	<title>SlamDunk</title>
	<meta charset="utf-8">
	<meta charset="utf-8">
	<meta content="IE=edge" http-equiv="X-UA-Compatible">
	<meta content="width=device-width, initial-scale=1" name="viewport">
	<meta content="SlamDunk website" name="description">
	<meta content="Tobias Neumann" name="author">
	<link href="https://netdna.bootstrapcdn.com/bootstrap/3.3.4/css/bootstrap.min.css" rel="stylesheet"><!--[if lte IE 8]><script src="js/ie/html5shiv.js"></script><![endif]-->
	<link href="css/main.css" rel="stylesheet"><!--[if lte IE 8]>
   <link rel="stylesheet" href="css/ie8.css" />
   <![endif]-->
	<!--[if lte IE 9]>
   <link rel="stylesheet" href="css/ie9.css" />
   <![endif]-->
</head>
<body id="top">
	<header id="header">
		<div class="header" style="position: relative; top: -100px;">
			<div class="navbar-static-top">
				<ul class="nav navbar-nav navbar-right">
					<li class="active">
						<a href="#">Home</a>
					</li>
					<li>
						<a href="docs.html">Docs</a>
					</li>
					<li>
						<a href="https://github.com/t-neumann/slamdunk">GitHub</a>
					</li>
				</ul>
			</div>
			<div class="jumbotron" style="position: relative; top: 100px;">
				<img src="images/slamdunk_logo_light.png" style="margin-bottom:100px;" width="400">
				<h3><b>Streamlining SLAM-Seq analysis with ultra-high sensitivity</b></h3>
				SlamDunk is a novel, fully automated software tool for automated, robust, scalable and reproducible SLAM-Seq data analysis.<br/>
				Diagnostic plotting features and our MultiQC plugin will make your SLAM-seq data ready for immediate QA and interpretation.<br/>  
				
				<div class="row" style="margin-top:30px;">
					<div class="col-md-3">
						<a class="btn btn-block btn-lg btn-primary" href="https://www.github.com/t-neumann/slamdunk"><i class="fa fa-github fa-lg"></i> GitHub</a>
					</div>
					<div class="col-md-3">
						<a class="btn btn-block btn-lg btn-primary" href="https://pypi.python.org/pypi/slamdunk"><i class="icon-python fa-lg"></i> PyPI</a>
					</div>
					<div class="col-md-3">
						<a class="btn btn-block btn-lg btn-primary" href="https://hub.docker.com/r/tobneu/slamdunk"><i class="fa fa-docker"></i> Docker</a>
					</div>
					<div class="col-md-3">
						<a class="btn btn-block btn-lg btn-primary" href="docs.html"><i class="fa fa-book fa-lg"></i> Docs</a>
					</div>
				</div>
			</div>
		</div>
	</header><!-- Intro -->
	<section class="wrapper" id="intro">
		<div class="inner">
			<header class="major">
				<h2>SLAM-Seq</h2>
			</header>
			<section>
				<h4><b>SLAM-Seq</b></h4>
				<p>SLAMSeq is a novel sequencing protocol to measure transcriptional output and transcript half-lifes.</p>
				<hr>
			</section>
		</div>
	</section><!-- Quickstart -->
	<section class="wrapper" id="quickstart">
		<div class="inner">
			<header class="major">
				<h2>Quickstart</h2>
			</header>
			<section>
				<h4><b>Install</b></h4>
				<div class="code bash highlight-python">
					<div class="highlight">
						<pre>pip install slamdunk</pre>
					</div>
				</div>
				<hr class="docutils">
			</section>
			<section>
				<h4><b>Run SlamDunk</b></h4>
				<div class="code bash highlight-python">
					<div class="highlight">
						<pre>slamdunk all -r &lt;reference fasta&gt; -b &lt;bed file&gt; -o &lt;output directory&gt; -5 12 -n 100
             -t &lt;threads&gt; -m -rl &lt;maximum read length&gt; --skip-sam files [files ...]
</pre>
					</div>
				</div>
				<p>This runs <i>slamdunk</i> with its default parameters.</p>
				<table border="1" class="docutils">
					<colgroup>
						<col width="6%">
						<col width="94%">
					</colgroup>
					<thead valign="bottom">
						<tr class="row-odd">
							<th class="head">Parameter</th>
							<th class="head">Description</th>
						</tr>
					</thead>
					<tbody valign="top">
						<tr class="row-even">
							<td><strong>-r</strong></td>
							<td>The reference fasta file.</td>
						</tr>
						<tr class="row-odd">
							<td><strong>-b</strong></td>
							<td>BED-file containing coordinates for 3’ UTRs.</td>
						</tr>
						<tr class="row-even">
							<td><strong>-o</strong></td>
							<td>The output directory where all output files of this dunk will be placed.</td>
						</tr>
						<tr class="row-odd">
							<td><strong>-t</strong></td>
							<td>The number of threads to use for this dunk. NextGenMap runs multi-threaded, so it is recommended to use more threads than available samples (default: 1).</td>
						</tr>
						<tr class="row-even">
							<td><strong>files</strong></td>
							<td>
								Samplesheet (see <a class="reference internal" href="docs.html#sample-file"><em>Sample file format</em></a> ) or a list of all sample BAM/FASTA(gz)/FASTQ(gz) files (wildcard * accepted).
							</td>
						</tr>
					</tbody>
				</table>
				<h4><b>Relevant results:</b></h4>
				<ul class="simple">
					<li>Folder <em>count</em> : Tab-separated <em>tcount</em> files containing the SLAM-Seq statistics per UTR (see <a class="reference internal" href="docs.html#tcount-file"><em>Tcount file format</em></a> )
					</li>
					<li>Folder <em>filter</em> : BAM-files with the final mapped reads for visualization and further processing</li>
				</ul>
				<p></p>
				<hr class="docutils">
			</section>
			<section>
				<h4><b>Check results</b></h4>If you want to have a very quick sanity check whether your desired conversion rates have been achieved, use this <em>alleyoop</em> command to plot the UTR conversion rates in your sample:
				<p></p>
				<div class="code bash highlight-python">
					<div class="highlight">
						<pre>alleyoop utrrates -o &lt;output directory&gt; -r &lt;reference fasta&gt; -t &lt;threads&gt;
                  -b &lt;bed file&gt; -l &lt;maximum read length&gt; bam [bam ...]
</pre>
					</div>
				</div>
				<table border="1" class="docutils">
					<colgroup>
						<col width="5%">
						<col width="95%">
					</colgroup>
					<thead valign="bottom">
						<tr class="row-odd">
							<th class="head">Parameter</th>
							<th class="head">Description</th>
						</tr>
					</thead>
					<tbody valign="top">
						<tr class="row-even">
							<td><strong>-o</strong></td>
							<td>The output directory where the plots will be placed.</td>
						</tr>
						<tr class="row-odd">
							<td><strong>-r</strong></td>
							<td>The reference fasta file.</td>
						</tr>
						<tr class="row-even">
							<td><strong>-t</strong></td>
							<td>The number of threads to use. All tools run single-threaded, so it is recommended to use as many threads as available samples.</td>
						</tr>
						<tr class="row-odd">
							<td><strong>-b</strong></td>
							<td>BED-file containing coordinates for 3’ UTRs.</td>
						</tr>
						<tr class="row-even">
							<td><strong>-l</strong></td>
							<td>Maximum read length in all samples.</td>
						</tr>
						<tr class="row-odd">
							<td><strong>bam</strong></td>
							<td>BAM file(s) containing the final filtered reads from the <em>filter</em> folder (wildcard * accepted).</td>
						</tr>
					</tbody>
				</table>
			</section>
		</div>
	</section><!-- Slamdunk description -->
	<section class="wrapper" id="slamdunk">
		<div class="inner alt">
			<header class="major">
				<h2>SlamDunk</h2>
				<div class="row" id="docstart">
					<div class="col-sm-6">
						<p align="left">SlamDunk consists of 4 core modules to process SLAM-Seq data:</p>
						<ul align="left">
							<li><b>map</b></li>
							<li style="list-style: none; display: inline">
								<ul>
									<li>Map reads to genome</li>
								</ul>
							</li>
							<li><b>filter</b></li>
							<li style="list-style: none; display: inline">
								<ul>
									<li>Filter alignments for low quality</li>
								</ul>
							</li>
							<li><b>snp</b></li>
							<li style="list-style: none; display: inline">
								<ul>
									<li>Call variants on data to filter false-positives</li>
								</ul>
							</li>
							<li><b>count</b></li>
							<li style="list-style: none; display: inline">
								<ul>
									<li>Quantify conversion rate per 3'UTR</li>
								</ul>
							</li>
						</ul>
					</div>
					<div class="col-sm-6" id="toc_column"><img src="images/slamdunk_flow.png" style="margin-bottom:100px;" width="400"></div>
				</div>
			</header>
			<section class="spotlight">
				<div class="image"><img alt="" src="images/ngm.png"></div>
				<div class="content">
					<h3>T&gt;C aware alignment</h3>
					<p>Using adapted scoring schemes for <a href="http://cibiv.github.io/NextGenMap/">NextGenMap</a> - a highly sensitive and fast read mapping program - we are able confidently map reads with any reasonable number of T&gt;C conversions to the genome.</p>
				</div>
			</section>
			<section class="spotlight">
				<div class="image"><img alt="" src="images/multimappers.png"></div>
				<div class="content">
					<h3>Multimapper reconciliation</h3>
					<p>We devised a reference guided approach to utilize multimapper information in 3'UTR regions of low complexity using efficient data structures.</p>
				</div>
			</section>
			<section class="spotlight">
				<div class="image"><img alt="" src="images/variant.png"></div>
				<div class="content">
					<h3>Variant exclusion</h3>
					<p>Using <a href="http://dkoboldt.github.io/varscan/">Varscan2</a> we can separate true T&gt;C conversions from false-positives caused by variants.</p>
				</div>
			</section>
			<section class="spotlight">
				<div class="image"><img alt="" src="images/conversionrates.png"></div>
				<div class="content">
					<h3>Conversion rate quantification</h3>
					<p>We devised a conversion counting method normalizing for T-content and coverage in individual 3'UTRs for unbiased comparison of transcripts.</p>
				</div>
			</section>
		</div>
	</section><!-- Installation -->
	<section class="wrapper" id="installation">
		<div class="inner">
			<header class="major">
				<h2>Installation</h2>
			</header>
			<section>
				<h3><b>Requirements</b></h3>
				<h4><b>SlamDunk</b></h4>
				<div class="panel panel-default">
					<div class="panel-heading">
						<h3 class="panel-title" style="color: #999999;"><b><i>slamdunk</i></b> will fetch all dependencies <b>automatically</b> in its install phase. There is no need to have the requirements installed before installing <i>slamdunk</i>.</h3>
					</div>
				</div>
				<p></p>
				<p><i>slamdunk</i> depends on several python packages listed in the <i>requirements.txt</i> file.</p>
				<p>In addition, <i>slamdunk</i> uses the following external software for its analysis:</p>
				<ul>
					<li><b><a href="http://cibiv.github.io/NextGenMap/">NextGenMap</a></b></li>
					<li><b><a href="http://cibiv.github.io/NextGenMap/">Samtools</a></b> &ge; 1.3.1</li>
					<li><b><a href="http://dkoboldt.github.io/varscan/">VarScan2</a></b> (requires <b>Java</b>)</li>
					<li><b><a href="https://www.r-project.org//">R</a></b> 3.2.2</li>
				</ul>
				<p></p>
				<h4><b>Alleyoop / Splash</b></h4>
				<p>Both <i>alleyoop</i> and <i>splash</i> require <b><a href="https://www.r-project.org/">R (v3.2.2)</a></b>. All R package dependencies are resolved automatically from <a href="https://cran.r-project.org">CRAN</a> during installation.</p>
			</section>
			<section>
				<h4><b>Installation</b></h4>
				<div class="panel panel-primary" id="quick_install">
					<div class="btn-group btn-group-justified btn-group-xs install-switcher" role="group">
						<a class="btn active" href="#pip">pip</a> <a class="btn" href="#dev">development version</a> <a class="btn" href="#manual">manual</a>
					</div>
					<div class="panel-body">
						<div id="pip">
							<p>SlamDunk is hosted as Python package on the <a href="https://pypi.python.org/pypi">Python Package Index</a>. You can install it with a single command using <a href="https://pypi.python.org/pypi/pip">pip</a>.</p>
							<pre><code><span class="comment"># Root</span>
pip install slamdunk

<span class="comment"># Local user</span>
pip install --user slamdunk</code></pre>
						</div>
						<div id="dev" style="display:none;">
							<p>The latest source of SlamDunk is hosted in our <a href="https://github.com/t-neumann/slamdunk">GitHub repository</a>. You can install it directly from there using <a href="https://pypi.python.org/pypi/pip">pip</a>.</p>
							<pre><code><span class="comment"># Grab source and install</span>
pip install git+https://github.com/t-neumann/slamdunk.git</code></pre>
						</div>
						<div id="manual" style="display:none;">
							<p>If you want to install all modules and dependencies manually from source, we have all install scripts in place.</p>
							<p><b>Note:</b> The recommended way to install python package dependencies is still using <a href="https://pypi.python.org/pypi/pip">pip</a>.</p>
							<pre><code>git clone https://github.com/t-neumann/slamdunk.git
cd slamdunk

<span class="comment"># Root</span>
pip install -r requirements.txt

<span class="comment"># Local user</span>
pip install --user -r requirements.txt

cd slamdunk/contrib

<span class="comment"># Build dependencies</span>
./build-ngm.sh
./build-varscan.sh
./build-samtools.sh
./build-rnaseqreadsimulator.sh

cd slamdunk/bin

<span class="comment"># Run from directory</span>
./slamdunk --help

<span class="comment"># Put it in your $PATH to run it from anywherel</span>
export PATH=$(pwd):$PATH

slamdunk --help</code></pre>
						</div>
					</div>
				</div>
				<p>For maximum convenience, you might also consider running slamdunk using our <a href="https://hub.docker.com/r/tobneu/slamdunk/">Docker</a> image.</p>
			</section>
		</div>
	</section><!-- Intro -->
	<section class="wrapper" id="multiqc">
		<div class="inner">
			<header class="major">
				<h2>MultiQC</h2>
			</header>
			<section>
				<div id="multiqc-description">
					<p><b><a href="http://multiqc.info">MultiQC</a></b> is a popular tool to aggregate results from bioinformatics analyses across many samples into a single report.</p>
					<p></p>
					<p>We implemented a MultiQC module which was released with <b>MultiQC v0.9</b> to support integration of SlamDunk QC plots and statistics into your MultiQC reports.</p>
					<p>Currently we support integration of the <em>summary</em>, <em>rates</em>, <em>utrrates</em>, <em>tcperreadpos</em> and <em>tcperutrpos</em> modules.</p>
					<div class="row">
						<div class="col-md-8">
							<p>You can download this report and / or the logs used to generate it, to try running MultiQC yourself.</p>
						</div>
						<div class="col-md-4">
							<div class="btn-group pull-right" role="group">
								<a class="btn btn-default download-btn" href="multiqc_example/slamdunk_report.zip"><span aria-hidden="true" class="fa fa-file-text-o"></span> Download report</a> <a class="btn btn-default download-btn" href="multiqc_example/slamdunk_logs.zip"><span aria-hidden="true" class="fa fa-files-o"></span> Download logs</a>
							</div>
						</div>
					</div>
				</div>
			</section>
			<section>
				<div class="hidden-xs hidden-sm" id="iframe_browser">
					<div id="iframe_browser_header">
						<div id="iframe_browser_buttons">
							<span></span><span></span><span></span>
						</div><span id="iframe_browser_title">MultiQC Example Reports</span>
						<ul id="iframe_browser_tabs">
							<li class="active">
								<a href="multiqc_example/multiqc_report.html">SlamDunk</a>
							</li>
						</ul>
					</div><iframe src="multiqc_example/multiqc_report.html"></iframe>
				</div>
				<p>Many thanks to <a href="https://github.com/ewels">Phil Ewels</a> for his support!</p>
			</section>
		</div>
	</section><!-- Authors -->
	<section class="wrapper" id="two">
		<div class="inner alt">
			<section class="spotlight">
				<div class="content">
					<h3>Authors and Contributors</h3>
					<p>SlamDunk is developed by <a href="mailto:tobias.neumann.at@gmail.com" target="_blank">Tobias Neumann</a> and <a href="mailto:philipp.rescheneder@univie.ac.at" target="_blank">Philipp Rescheneder</a>.</p>
				</div>
			</section>
		</div>
	</section><!--Footer -->
	<footer id="footer">
		<div class="row">
			<div class="col-md-6">
				<div class="icons" style="margin-left: 2em">
					<a href="https://github.com/t-neumann" target="_blank"><i class="fa fa-github"></i> t-neumann</a> | <a href="https://github.com/philres" target="_blank"><i class="fa fa-github"></i> philres</a>
				</div>
			</div>
			<div class="col-md-6">
				<p class="copyright">&copy; Tobias Neumann.<br>
				Page template: <a href="http://html5up.net">HTML5 UP</a></p>
			</div>
		</div>
	</footer><!-- Scripts -->
	<script>
	                 (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
	                 (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
	                 m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
	                 })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
	                 
	                 ga('create', 'UA-89369264-1', 'auto');
	                 ga('send', 'pageview');
	                 
	</script> 
	<script src="js/jquery.min.js">
	</script> 
	<script src="js/jquery.scrolly.min.js">
	</script> 
	<script src="js/skel.min.js">
	</script> 
	<script src="js/util.js">
	</script> <!--[if lte IE 8]><script src="js/ie/respond.min.js"></script><![endif]-->
	 
	<script src="js/main.js">
	</script> 
	<script src="js/slamdunk.js">
	</script>
</body>
</html>