
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    
    <title>Documentation: MultiQC</title>
    <meta name="description" content="">
    <meta name="author" content="">
    <link rel="apple-touch-icon" sizes="57x57" href="../apple-icon-57x57.png">
    <link rel="apple-touch-icon" sizes="60x60" href="../apple-icon-60x60.png">
    <link rel="apple-touch-icon" sizes="72x72" href="../apple-icon-72x72.png">
    <link rel="apple-touch-icon" sizes="76x76" href="../apple-icon-76x76.png">
    <link rel="apple-touch-icon" sizes="114x114" href="../apple-icon-114x114.png">
    <link rel="apple-touch-icon" sizes="120x120" href="../apple-icon-120x120.png">
    <link rel="apple-touch-icon" sizes="144x144" href="../apple-icon-144x144.png">
    <link rel="apple-touch-icon" sizes="152x152" href="../apple-icon-152x152.png">
    <link rel="apple-touch-icon" sizes="180x180" href="../apple-icon-180x180.png">
    <link rel="icon" type="image/png" sizes="192x192"  href="../android-icon-192x192.png">
    <link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
    <link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
    <link rel="manifest" href="../manifest.json">
    <meta name="msapplication-TileColor" content="#ffffff">
    <meta name="msapplication-TileImage" content="../ms-icon-144x144.png">
    <meta name="theme-color" content="#ffffff">

    <link href="https://netdna.bootstrapcdn.com/bootstrap/3.3.4/css/bootstrap.min.css" rel="stylesheet">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
    <link rel="stylesheet" href="../fonts/font-mfizz.css">
    <link rel="stylesheet" href="../css/code_highlighting/github-light.css">
    <link rel="stylesheet" href="../css/main.css">

    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
  </head>
  
  

  <body id="mqc_docs" data-spy="scroll" data-target="#toc" data-offset="100">
  
  <div class="header">
      
      <header id="header">
         <div class="jumbotron" style="background-color:transparent;">
         <img src="images/slamdunk_logo_light.png" width="400" style="margin-bottom:100px;">
         <h3>
            <b>Streamlining SLAMSeq analysis with ultra-high sensitivity</b>
         </h3>
         Slamdunk analyzes SLAMSeq data.
         <div class="row" style="margin-top:30px;">
            <div class="col-md-3">
               <a class="btn btn-block btn-lg btn-primary" href="https://www.github.com/t-neumann/slamdunk">
               <i class="fa fa-github fa-lg"></i> GitHub
               </a>
            </div>
            <div class="col-md-3">
               <a class="btn btn-block btn-lg btn-primary" href="https://pypi.python.org/pypi/slamdunk">
               <i class="icon-python fa-lg"></i> PyPI
               </a>
            </div>
            <div class="col-md-3">
               <a class="btn btn-block btn-lg btn-primary" href="https://hub.docker.com/r/tobneu/slamdunk-full">
               <i class="fa fa-docker"></i> Docker
               </a>
            </div>
            <div class="col-md-3">
               <a class="btn btn-block btn-lg btn-primary" href="https://slamdunk.readthedocs.io">
               <i class="fa fa-book fa-lg"></i> Docs
               </a>
            </div>
         </div>
      </header>
      </div>
      
    <div class="container docs-container">
      <div class="row">
        <div class="col-sm-9">
          <div class="docs_section">
<h1 class="section-header" id="using-multiqc"><a href="#using-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using MultiQC</h1>
<div class="docs_block" id="installation.md"><h1 id="installing-multiqc"><a href="#installing-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installing MultiQC</h1>
<h2 id="installing-python"><a href="#installing-python" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installing Python</h2>
<p>To see if you have python installed, run <code>python --version</code> on the command line.
If you see version 2.7+, 3.4+ or 3.5+ then you can skip this step.</p>
<p>We recommend using virtual environments to manage your Python installation.
Our favourite is Anaconda, a cross-platform tool to manage Python environments.
You can installation instructions for Anaconda
<a href="http://conda.pydata.org/docs/install/quick.html">here</a>.</p>
<p>Once Anaconda is installed, you can create an environment with the
following commands:</p>
<pre><code>conda create --name py2.7 python=2.7
source activate py2.7
# Windows: activate py2.7</code></pre>
<p>You'll want to add the <code>source activate py2.7</code> line to your <code>.bashrc</code> file so
that the environment is loaded every time you load the terminal.</p>
<h2 id="installing-with-conda"><a href="#installing-with-conda" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installing with conda</h2>
<p>If you're using <code>conda</code> as described above, you can install MultiQC from the <code>bioconda</code>
channel as follows:</p>
<pre><code>conda install -c bioconda multiqc</code></pre>
<h2 id="installation-with-pip"><a href="#installation-with-pip" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installation with pip</h2>
<p>This is the easiest way to install MultiQC. <code>pip</code> is the package manager for
the Python Package Manager. It comes bundled with recent versions of Python,
otherwise you can find installation instructions <a href="http://pip.readthedocs.org/en/stable/installing/">here</a>.</p>
<p>You can now install MultiQC from
<a href="https://pypi.python.org/pypi/multiqc">PyPI</a> as follows:</p>
<pre><code>pip install multiqc</code></pre>
<p>If you would like the development version, the command is:</p>
<pre><code>pip install git+https://github.com/ewels/MultiQC.git</code></pre>
<p>Note that if you have problems with read-only directories, you can install to
your home directory with the <code>--user</code> parameter (though it's probably better
to use virtual environments, as described above).</p>
<pre><code>pip install --user multiqc</code></pre>
<h2 id="manual-installation"><a href="#manual-installation" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Manual installation</h2>
<p>If you'd rather not use either of these tools, you can clone the code and install the code yourself:</p>
<pre><code>git clone https://github.com/ewels/MultiQC.git
cd MultiQC
python setup.py install</code></pre>
<p><code>git</code> not installed? No problem - just download the flat files:</p>
<pre><code>curl -LOk https://github.com/ewels/MultiQC/archive/master.zip
unzip master.zip
cd MultiQC-master
python setup.py install</code></pre>
<h2 id="updating-multiqc"><a href="#updating-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Updating MultiQC</h2>
<p>You can update MultiQC from <a href="https://pypi.python.org/pypi/multiqc">PyPI</a>
at any time by running the following command:</p>
<pre><code>pip --update multiqc</code></pre>
<p>To update the development version, use:</p>
<pre><code>pip install --force git+https://github.com/ewels/MultiQC.git</code></pre>
<p>If you cloned the <code>git</code> repo, just pull the latest changes and install:</p>
<pre><code>cd MultiQC
git pull
python setup.py install</code></pre>
<p>If you downloaded the flat files, just repeat the installation procedure.</p>
<h2 id="installing-as-an-environment-module"><a href="#installing-as-an-environment-module" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installing as an environment module</h2>
<p>Many people using MultiQC will be working on a HPC environment.
Every server / cluster is different, and you're probably best off asking
your friendly sysadmin to install MultiQC for you. However, with that
in mind, here are a few general tips for installing MultiQC into an
environment module system:</p>
<p>MultiQC comes in two parts - the <code>multiqc</code> python package and the
<code>multiqc</code> executable script. The former must be available in <code>$PYTHONPATH</code>
and the script must be available on the <code>$PATH</code>.</p>
<p>A typical installation procedure with an environment module Python install
might look like this: <em>(Note that <code>$PYTHONPATH</code> must be defined before <code>pip</code> installation.)</em></p>
<pre><code>VERSION=0.7
INST=/path/to/software/multiqc/$VERSION
module load python/2.7.6
mkdir $INST
export PYTHONPATH=$INST/lib/python2.7/site-packages
pip install --install-option="--prefix=$INST" multiqc</code></pre>
<p>Once installed, you'll need to create an environment module file.
Again, these vary between systems a lot, but here's an example:</p>
<pre><code class="language-bash">#%Module1.0#####################################################################
##
## MultiQC
##

set components [ file split [ module-info name ] ]
set version [ lindex $components 1 ]
set modroot /path/to/software/multiqc/$version

proc ModulesHelp { } {
    global version modroot
    puts stderr "\tMultiQC - use MultiQC $version"
    puts stderr "\n\tVersion $version\n"
}
module-whatis   "Loads MultiQC environment."

# load required modules
module load python/2.7.6

# only one version at a time
conflict multiqc

# Make the directories available
prepend-path    PATH        $modroot/bin
prepend-path    PYTHONPATH  $modroot/lib/python2.7/site-packages</code></pre></div><div class="docs_block" id="usage.md"><h1 id="running-multiqc"><a href="#running-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Running MultiQC</h1>
<p>Once installed, just go to your analysis directory and run <code>multiqc</code>, followed
by a list of directories to search. At it's simplest, this can just be <code>.</code>
(the current working directory):</p>
<pre><code>multiqc .</code></pre>
<p>That's it! MultiQC will scan the specified directories and produce a report
based on details found in any log files that it recognises.</p>
<p>See <a href="http://multiqc.info/docs/#using-multiqc-reports">Using MultiQC Reports</a> for more information about how
to use the generated report.</p>
<p>For a description of all command line parameters, run <code>multiqc --help</code>.</p>
<h2 id="choosing-where-to-scan"><a href="#choosing-where-to-scan" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Choosing where to scan</h2>
<p>You can supply MultiQC with as many directories or files as you like. Above,
we supply <code>.</code> - just the current directory, but all of these would work too:</p>
<pre><code>multiqc data/
multiqc data/ ../proj_one/analysis/ /tmp/results
multiqc data/*_fastqc.zip
multiqc data/sample_1*</code></pre>
<p>You can also ignore files using the <code>-x</code>/<code>--ignore</code> flag (can be specified multiple
times). This takes a string which it matches using glob expansion to filenames,
directory names and entire paths:</p>
<pre><code>multiqc . --ignore *_R2*
multiqc . --ignore run_two/
multiqc . --ignore */run_three/*/fastqc/*_R2.zip</code></pre>
<p>Finally, you can supply a file containing a list of file paths, one per row.
MultiQC only search the listed files.</p>
<pre><code>multiqc --file-list my_file_list.txt</code></pre>
<h2 id="renaming-reports"><a href="#renaming-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Renaming reports</h2>
<p>The report is called <code>multiqc_report.html</code> by default. Tab-delimited data files
are created in <code>multiqc_data/</code>, containing additional information.
You can use a custom name for the report with the <code>-n</code>/<code>--name</code> parameter, or instruct
MultiQC to create them in a subdirectory using the <code>-o</code>/<code>-outdir</code> parameter.</p>
<p>Note that different MultiQC templates may have different defaults.</p>
<h2 id="overwriting-existing-reports"><a href="#overwriting-existing-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Overwriting existing reports</h2>
<p>It's quite common to repeatedly create new reports as new analysis results
are generated. Instead of manually deleting old reports, you can just specify
the <code>-f</code> parameter and MultiQC will overwrite any conflicting report filenames.</p>
<h2 id="sample-names-prefixed-with-directories"><a href="#sample-names-prefixed-with-directories" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Sample names prefixed with directories</h2>
<p>Sometimes, the same samples may be processed in different ways. If MultiQC
finds log files with the same sample name, the previous data will be overwritten
(this can be inspected by running MultiQC with <code>-v</code>/<code>--verbose</code>).</p>
<p>To avoid this, run MultiQC with the <code>-d</code>/<code>--dirs</code> parameter. This will prefix every
sample name with the directory path that the log file was found within. As
such, sample names will no longer be unique, and data will not be overwritten.</p>
<p>By default, <code>--dirs</code> will prepend the entire path to each sample name. You can choose
which directories are added with the <code>-dd</code>/<code>--dirs-depth</code> parameter. Set to a positive
integer to use that many directories at the end of the path. A negative integer takes
directories from the start of the path.</p>
<p>For example:</p>
<pre><code>$ multiqc -d .
# analysis_1 | results | type | sample_1 | file.log
# analysis_2 | results | type | sample_2 | file.log
# analysis_3 | results | type | sample_3 | file.log

$ multiqc -d -dd 1 .
# sample_1 | file.log
# sample_2 | file.log
# sample_3 | file.log

$ multiqc -d -dd -1 .
# analysis_1 | file.log
# analysis_2 | file.log
# analysis_3 | file.log</code></pre>
<h2 id="using-different-templates"><a href="#using-different-templates" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using different templates</h2>
<p>MultiQC is built around a templating system. You can produce reports with
different styling by using the <code>-t</code>/<code>--template</code> option. The available templates
are listed with <code>multiqc --help</code>.</p>
<p>If you're interested in creating your own custom template, see the
<a href="http://multiqc.info/docs/#writing-new-templates">writing new templates</a> section.</p>
<h2 id="pdf-reports"><a href="#pdf-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>PDF Reports</h2>
<p>Whilst HTML is definitely the format of choice for MultiQC reports due to
the interactive features that it can offer, PDF files are an integral part
of some people's workflows. To try to accommodate this, MultiQC has a
<code>--pdf</code> command line flag which will try to create a PDF report for you.</p>
<p>To do this, MultiQC uses the <code>simple</code> template. This uses flat plots,
has no navigation or toolbar and strips out all JavaScript. The resulting
HTML report is pretty basic, but this simplicity is helpful when generating
PDFs.</p>
<p>Once the report is generated MultiQC attempts to call <a href="http://pandoc.org/">Pandoc</a>,
a command line tool able to convert documents between different file formats.
<strong>You must have Pandoc already installed for this to work</strong>. If you don't have
Pandoc installed, you will get an error message that looks like this:</p>
<pre><code>Error creating PDF - pandoc not found. Is it installed? http://pandoc.org/</code></pre>
<p>Please note that Pandoc is a complex tool and uses LaTeX / XeLaTeX for PDF
generation. Please make sure that you have the latest version of Pandoc and
that it can successfully convert basic HTML files to PDF before reporting
and errors. Also note that not all plots have flat image equivalents, so
some will be missing (at time of writing: FastQC sequence content plot,
beeswarm dot plots, heatmaps).</p>
<h2 id="printing-to-stdout"><a href="#printing-to-stdout" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Printing to stdout</h2>
<p>If you would like to generate MultiQC reports on the fly, you can print the
output to standard out by specifying <code>-n stdout</code>. Note that the data directory
will not be generated and the template used must create stand-alone HTML reports.</p>
<h2 id="parsed-data-directory"><a href="#parsed-data-directory" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Parsed data directory</h2>
<p>By default, MultiQC creates a directory alongside the report containing
tab-delimited files with the parsed data. This is useful for downstream
processing, especially if you're running MultiQC with very large numbers
of samples.</p>
<p>Typically, these files are tab-delimited tables. However, you can get <code>JSON</code>
or <code>YAML</code> output for easier downstream parsing by specifying <code>-k</code>/<code>--data-format</code>
on the command line or <code>data_format</code> in your configuration file.</p>
<p>You can also choose whether to produce the data by specifying either the
<code>--data-dir</code> or <code>--no-data-dir</code> command line flags or the <code>make_data_dir</code>
variable in your configuration file. Note that the data directory
is never produced when printing the MultiQC report to <code>stdout</code>.</p>
<p>To zip the data directory, use the <code>-z</code>/<code>--zip-data-dir</code> flag.</p>
<h2 id="exporting-plots"><a href="#exporting-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Exporting Plots</h2>
<p>In addition to the HTML report, it's also possible to get MultiQC to save
plots as stand alone files. You can do this with the <code>-p</code>/<code>--export</code> command
line flag. By default, plots will be saved in a directory called <code>multiqc_plots</code>
as <code>.png</code>, <code>.svg</code> and <code>.pdf</code> files.</p>
<p>You can instruct MultiQC to always do this by setting the <code>export_plots</code> config
option to <code>true</code>, though note that this will add a few seconds on to execution time.
The <code>plots_dir_name</code> changes the default directory name for plots and the
<code>export_plot_formats</code> specifies what file formats should be created (must be
supported by MatPlotLib).</p>
<p>Note that not all plot types are yet supported, so you may find some plots are
missing.</p>
<blockquote>
<p>Note: You can always save static image versions of plots from within
MultiQC reports, using the <a href="http://multiqc.info/docs/#export">Export toolbox</a> in the side bar.</p>
</blockquote>
<h2 id="choosing-which-modules-to-run"><a href="#choosing-which-modules-to-run" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Choosing which modules to run</h2>
<p>Sometimes, it's desirable to choose which MultiQC modules run. This could be
because you're only interested in one type of output and want to keep the
reports small. Or perhaps the output from one module is misleading in your
situation.</p>
<p>You can do this by using <code>-m</code>/<code>--modules</code> to explicitly define which modules
you want to run. Alternatively, use <code>-e</code>/<code>--exclude</code> to run all modules
except those listed.</p></div><div class="docs_block" id="reports.md"><h1 id="using-multiqc-reports"><a href="#using-multiqc-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using MultiQC Reports</h1>
<p>Once MultiQC has finished, you should have a HTML report file called
<code>multiqc_report.html</code> (or something similar, depending on how you ran MultiQC).
You can launch this report with <code>open multiqc_report.html</code> on the command
line, or double clicking the file in a file browser.</p>
<h2 id="browser-compatibility"><a href="#browser-compatibility" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Browser compatibility</h2>
<p>MultiQC reports should work in any modern browser. They have been tested
using OSX Chrome, Firefox and Safari. If you find any report bugs, please
report them as a <a href="https://github.com/ewels/MultiQC/issues">GitHub issue</a>.</p>
<h2 id="report-layout"><a href="#report-layout" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Report layout</h2>
<p>MultiQC reports have three main page sections:</p>
<ul>
<li>The navigation menu (left side)
<ul>
<li>Links to the different module sections in the report</li>
<li>Click the logo to go to the top of the page</li>
</ul></li>
<li>The toolbox (right side)
<ul>
<li>Contains various tools to modify the report data (see below)</li>
</ul></li>
<li>The report (middle)
<ul>
<li>This is what you came here for, the data!</li>
</ul></li>
</ul>
<p>Note that if you're viewing the report on a mobile device / small window,
the content will be reformatted to fit the screen.</p>
<h2 id="general-statistics-table"><a href="#general-statistics-table" class="header-link"><span class="glyphicon glyphicon-link"></span></a>General Statistics table</h2>
<p>At the top of every MultiQC report is the 'General Statistics' table.
This shows an overview of key values, taken from all modules. The aim
of the table is to bring together stats for each sample from across the
analysis so that you can see it in one place.</p>
<p>Hovering over column headers will show a longer description, including which
module produced the data. Clicking a header will sort the table by that value.
Clicking it again will change the sort direction. You can shift-click multiple
headers to sort by multiple columns.</p>
<p><img src="images/genstats_sort.png" alt="sort column" /></p>
<p>Above the table there is a button called 'Configure Columns'. Clicking this
will launch a modal window with more detailed information about each column,
plus options to show/hide and change the order of columns.</p>
<p><img src="images/genstats_config_cols.png" alt="configure columns" /></p>
<h2 id="plots"><a href="#plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Plots</h2>
<p>MultiQC modules can take plot more extensive data in the sections below
the general statistics table.</p>
<h3 id="interactive-plots"><a href="#interactive-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Interactive plots</h3>
<p>Plots in MultiQC reports are usually interactive, using the
<a href="http://www.highcharts.com/">HighCharts</a> JavaScript library.</p>
<p>You can hover the mouse over data to see a tooltip with more information
about that dataset. Clicking and dragging on line graphs will zoom into that area.</p>
<p><img src="images/plot_zoom.png" alt="plot zoom" /></p>
<p>To reset the zoom, use the button in the top right:</p>
<p><img src="images/plot_reset_zoom.png" alt="reset zoom" /></p>
<p>Plots have a grey bar along their base; clicking and dragging this will
resize the plot's height:</p>
<p><img src="images/plot_height.png" alt="plot zoom" /></p>
<p>You can force reports to use interactive plots instead of flat by specifying
the <code>--interactive</code> command line option (see below).</p>
<h3 id="flat-plots"><a href="#flat-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Flat plots</h3>
<p>Reports with large numbers of samples may contain flat plots. These are
rendered when the MultiQC report is generated using
<a href="http://matplotlib.org/">MatPlotLib</a> and are non-interactive (flat)
images within the report. The reason for generating these is that large
sample numbers can make MultiQC reports very data-intensive and unresponsive
(crashing people's browsers in extreme cases). Plotting data in flat images
is scalable to any number of samples, however.</p>
<p>Flat plots in MultiQC have been designed to look as similar to their interactive
versions as possible. They are also copied to <code>multiqc_data/multiqc_plots</code></p>
<p>You can force reports to use flat plots with the <code>--flat</code> command line option.</p>
<p>See the <a href="http://multiqc.info/docs/#large-sample-numbers">Large sample numbers</a> section of the
<em>Configuring MultiQC</em> docs for more on how to customise the flat / interactive
plot behaviour.</p>
<h3 id="exporting-plots-1"><a href="#exporting-plots-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Exporting plots</h3>
<p>If you want to use the plot elsewhere (<em>eg.</em> in a presentation or paper),
you can export it in a range of formats. Just click the menu button in
the top right of the plot:</p>
<p><img src="images/plot_export.png" alt="plot zoom" /></p>
<p>This opens the MultiQC Toolbox <em>Export Plots</em> panel with the current plot
selected. You have a range of export options here. When deciding on output
format bear in mind that SVG is a vector format, so can be edited in tools
such as <a href="http://www.adobe.com/products/illustrator.html">Adobe Illustrator</a>
or the free tool <a href="https://inkscape.org/en/">Inkscape</a>. This makes it ideal
for use in publications and manual customisation / annotation. The
<em>Plot scaling</em> option changes how large the labels are relative to the plot.</p>
<h3 id="dynamic-plots"><a href="#dynamic-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Dynamic plots</h3>
<p>Some plots have buttons above them which allow you to change the data
that they show or their axis. For example, many bar plots have the option
to show the data as percentages instead of counts:</p>
<p><img src="images/plot_percentage_button.png" alt="percentage button" /></p>
<h2 id="toolbox"><a href="#toolbox" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Toolbox</h2>
<p>MultiQC reports come with a 'toolbox', accessible by clicking the buttons
on the right hand side of the report:</p>
<p><img src="images/toolbox_buttons.png" alt="toolbox buttons" /></p>
<p>Active toolbox panels have their button highlighted with a blue outline.
You can hide the toolbox by clicking the open panel button a second time,
or pressing <code>Escape</code> on your keyboard.</p>
<h3 id="highlight-samples"><a href="#highlight-samples" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Highlight Samples</h3>
<p>If you run MultiQC plots with a lot of samples, plots can become very
data-heavy. This makes it difficult to find specific samples, or subsets
of samples.</p>
<p>To help with this, you can use the Highlight Samples tool to colour datasets
of interest. Simply enter some text which will match the samples you want to
highlight and press enter (or click the add button). If you like, you can also
customise the highlight colour.</p>
<p><img src="images/toolbox_highlight.png" alt="toolbox highlight" /></p>
<p>To make it easier to match groups of samples, you can use a regular expressions
by turning on 'Regex mode'. You can test regexes using a nice tool at
<a href="https://regex101.com/">regex101.com</a>. See a nice introduction to regexes
<a href="http://www.regular-expressions.info/quickstart.html">here</a>. Note that pattern
delimiters are not needed (use <code>pattern</code>, not <code>/pattern/</code>).</p>
<p>Here, we highlight any sample names that end in <code>_1</code>:</p>
<p><img src="images/toolbox_highlight_regex.png" alt="highligh regex" /></p>
<p>Note that a new button appears above the General Statistics table when samples
are highlighted, allowing you to sort the table according to highlights.</p>
<p>Search patterns can be changed after creation, just click to edit. To remove,
click the grey cross on the right hand side.</p>
<p>Searching for an empty string will match all samples.</p>
<h3 id="renaming-samples"><a href="#renaming-samples" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Renaming Samples</h3>
<p>Sample names are typically generated based on processed file names. These
file names are not always informative. To help with this, you can do a search
and replace within sample names. Here, we remove the <code>SRR1067</code> and <code>_1</code> parts
of the sample names, which are the same for all samples:</p>
<p><img src="images/toolbox_rename.png" alt="rename samples" /></p>
<p>Again, regular expressions can be used. See above for details. Note that
regex groups can be used - define a group match with parentheses and
use the matching value with <code>$1</code>, <code>$2</code> etc. For example - a search string
<code>SRR283(\d{3})</code> and replace string <code>$1_SRR283</code> would move the final three
digits of matching sample names to the start of the name.</p>
<p>Often, you may have a spreadsheet with filenames and informative sample
names. To avoid having to manually enter each name, you can paste from a
spreadsheet using the 'bulk import' tool:</p>
<p><img src="images/toolbox_bulk_rename.png" alt="bulk rename" /></p>
<h3 id="hiding-samples"><a href="#hiding-samples" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Hiding Samples</h3>
<p>Sometimes, you want to focus on a subset of samples. To temporarily
hide samples from the report, enter a search string as described above into
the 'Hide Samples' toolbox panel.</p>
<p>Here, we hide all samples with <code>_trimmed</code> in their sample name:
<em>(Note that plots will tell you how many samples have been hidden)</em></p>
<p><img src="images/toolbox_hide_samples.png" alt="hide samples" /></p>
<h3 id="export"><a href="#export" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Export</h3>
<p>This tool enables you to configure the size and characteristics of exported
plots, as well as allowing you to download some or all of the graphs with a
single click. Width and Height set the output size of the images, scale sets
how &quot;zoomed-in&quot; they should look (typically you want the plot to be more
zoomed for printing). See <a href="http://multiqc.info/docs/#exporting-plots">above</a> for more information
about this panel.</p>
<h3 id="save-settings"><a href="#save-settings" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Save Settings</h3>
<p>To avoid having to re-enter the same toolbox setup repeatedly, you can
save your settings using the 'Save Settings' panel. Just pick a name
and click save. To load, choose your set of settings and press load
(or delete). Loaded settings are applied on top of current settings.
All configs are saved in browser local storage - they do not travel
with the report and may not work in older browsers.</p></div><div class="docs_block" id="config.md"><h1 id="configuring-multiqc"><a href="#configuring-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Configuring MultiQC</h1>
<p>Whilst most MultiQC settings can be specified on the command line, MultiQC
is also able to parse system-wide and personal config files. At run time,
it collects the configuration settings from the following places in this order
(overwriting at each step if a conflicting config variable is found):</p>
<ol>
<li>Hardcoded defaults in MultiQC code</li>
<li>System-wide config in <code>&lt;installation_dir&gt;/multiqc_config.yaml</code>
<ul>
<li>Manual installations only, not <code>pip</code> or <code>conda</code></li>
</ul></li>
<li>User config in <code>~/.multiqc_config.yaml</code></li>
<li>Config file in the current working directory: <code>multiqc_config.yaml</code></li>
<li>Command line options</li>
</ol>
<p>You can find an example configuration file with the MultiQC source code, called
<a href="https://github.com/ewels/MultiQC/blob/master/multiqc_config_example.yaml"><code>multiqc_config.example.yaml</code></a>.
If you installed MultiQC with <code>pip</code> or <code>conda</code> you won't have this file locally,
but you can find it on GitHub:
<a href="https://github.com/ewels/MultiQC/blob/master/multiqc_config_example.yaml">github.com/ewels/MultiQC</a>.</p>
<h2 id="sample-name-cleaning"><a href="#sample-name-cleaning" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Sample name cleaning</h2>
<p>MultiQC typically generates sample names by taking the input or log file name,
and 'cleaning' it. To do this, it uses the <code>fn_clean_exts</code> settings and looks
for any matches. If it finds any matches, everything to the right is removed.
For example, consider the following config:</p>
<pre><code class="language-yaml">fn_clean_exts:
    - '.gz'
    - '.fastq'</code></pre>
<p>This would make the following sample names:</p>
<pre><code>mysample.fastq.gz  -&gt;  mysample
secondsample.fastq.gz_trimming_log.txt  -&gt;  secondsample
thirdsample.fastq_aligned.sam.gz  -&gt;  thirdsample</code></pre>
<p>There is also a config list called <code>fn_clean_trim</code> which just removes
strings if they are present at the start or end of the sample name.</p>
<p>Usually you don't want to overwrite the defaults (though you can).
Instead, add to the special variable names <code>extra_fn_clean_exts</code>
and <code>extra_fn_clean_trim</code>:</p>
<pre><code class="language-yaml">extra_fn_clean_exts:
    - '.myformat'
    - '_processedFile'
extra_fn_clean_trim:
    - '#'
    - '.myext'</code></pre>
<h3 id="other-search-types"><a href="#other-search-types" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Other search types</h3>
<p>File name cleaning can also take strings to remove (instead of removing with truncation).
Also regex strings can be supplied to match patterns and remove strings.</p>
<p>Consider the following:</p>
<pre><code class="language-yaml">extra_fn_clean_exts:
    - '.fastq'
    - type: 'replace'
      pattern: '.sorted'
    - type: 'regex'
      pattern: '^processed.'</code></pre>
<p>This would make the following sample names:</p>
<pre><code>mysample.fastq.gz  -&gt;  mysample
secondsample.sorted.deduplicated.fastq.gz_processed.txt  -&gt;  secondsample.deduplicated
processed.thirdsample.fastq_aligned.sam.gz  -&gt;  thirdsample</code></pre>
<h3 id="clashing-sample-names"><a href="#clashing-sample-names" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Clashing sample names</h3>
<p>This process of cleaning sample names can sometimes result in exact duplicates.
A duplicate sample name will overwrite previous results. Warnings showing these events
can be seen with verbose logging using the <code>--verbose</code>/<code>-v</code> flag, or in <code>multiqc_data/multiqc.log</code>.</p>
<p>Problems caused by this will typically be discovered be fewer results than expected. If you're ever
unsure about where the data from results within MultiQC reports come from, have a look at
<code>multiqc_data/multiqc_sources.txt</code>, which lists the path to the file used for every section
of the report.</p>
<h4>Directory names</h4>
<p>One scenario where clashing names can occur is when the same file is processed in different directories.
For example, if <code>sample_1.fastq</code> is processed with four sets of parameters in four different
directories, they will all have the same name - <code>sample_1</code>. Only the last will be shown.
If the directories are different, this can be avoided with the <code>--dirs</code>/<code>-d</code> flag.</p>
<p>For example, given the following files:</p>
<pre><code>├── analysis_1
│   └── sample_1.fastq.gz.aligned.log
├── analysis_2
│   └── sample_1.fastq.gz.aligned.log
└── analysis_3
    └── sample_1.fastq.gz.aligned.log</code></pre>
<p>Running <code>multiqc -d .</code> will give the following sample names:</p>
<pre><code>analysis_1 | sample_1
analysis_2 | sample_1
analysis_3 | sample_1</code></pre>
<h4>Filename truncation</h4>
<p>If the problem is with filename truncation, you can also use the <code>--fullnames</code>/<code>-s</code> flag,
which disables all sample name cleaning. For example:</p>
<pre><code>├── sample_1.fastq.gz.aligned.log
└── sample_1.fastq.gz.subsampled.fastq.gz.aligned.log</code></pre>
<p>Running <code>multiqc -s .</code> will give the following sample names:</p>
<pre><code>sample_1.fastq.gz.aligned.log
sample_1.fastq.gz.subsampled.fastq.gz.aligned.log</code></pre>
<p>You can turn off sample name cleaning permanently by setting
<code>fn_clean_sample_names</code> to <code>false</code> in your config file.</p>
<h2 id="module-search-patterns"><a href="#module-search-patterns" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Module search patterns</h2>
<p>Many bioinformatics tools have standard output formats, filenames and other
signatures. MultiQC uses these to find output; for example, the FastQC module
looks for files that end in <code>_fastqc.zip</code>.</p>
<p>This works well most of the time, until someone has an automated processing
pipeline that renames things. For this reason, as of version v0.3.2 of MultiQC,
the file search patterns are loaded as part of the main config. This means that
they can be overwritten in <code>&lt;installation_dir&gt;/multiqc_config.yaml</code> or
<code>~/.multiqc_config.yaml</code>. So if you always rename your <code>_fastqc.zip</code> files to
<code>_qccheck.zip</code>, MultiQC can still work.</p>
<p>To see the default search patterns, see the
<a href="https://github.com/ewels/MultiQC/blob/master/multiqc/utils/search_patterns.yaml"><code>search_patterns.yaml</code></a>
file. Copy the section for the program that you want to modify and paste this
into your config file. Make sure you make it part of a dictionary called <code>sp</code>
as follows:</p>
<pre><code class="language-yaml">sp:
    mqc_module:
        fn: _mysearch.txt</code></pre>
<p>Search patterns can specify a filename match (<code>fn</code>) or a file contents
match (<code>contents</code>).</p>
<h2 id="ignoring-files"><a href="#ignoring-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Ignoring Files</h2>
<p>MultiQC begins by indexing all of the files that you specified and building a list
of the ones it will use. You can specify files and directories to skip on the command
line using <code>-x</code>/<code>--ignore</code>, or for more permanent memory, with the following config file
options: <code>fn_ignore_files</code>, <code>fn_ignore_dirs</code> and <code>fn_ignore_paths</code> (the command line
option simply adds to all of these).</p>
<p>For example, given the following files:</p>
<pre><code>├── analysis_1
│   └── sample_1.fastq.gz.aligned.log
├── analysis_2
│   └── sample_1.fastq.gz.aligned.log
└── analysis_3
    └── sample_1.fastq.gz.aligned.log</code></pre>
<p>You could specify the following relevant config options:</p>
<pre><code class="language-yaml">fn_ignore_files:
    - '*.log'
fn_ignore_dirs:
    - 'analysis_1'
    - 'analysis_2'
fn_ignore_paths:
    - '*/analysis_*/sample_1*'</code></pre>
<p>Note that the searched file paths will usually be relative to the working
directory and can be highly variable, so you'll typically want to start patterns
with a <code>*</code> to match any preceding directory structure.</p>
<h2 id="large-sample-numbers"><a href="#large-sample-numbers" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Large sample numbers</h2>
<p>MultiQC has been written with the intention of being used for any number of samples.
This means that it <em>should</em> work well with 6 samples or 6000. Very large sample numbers
are becoming increasingly common, for example with single cell data.</p>
<p>Producing reports with data from many hundreds or thousands of samples provides some
challenges, both technically and also in terms of data visualisation and report usability.</p>
<h3 id="disabling-on-load-plotting"><a href="#disabling-on-load-plotting" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Disabling on-load plotting</h3>
<p>One problem with large reports is that the browser can hang when the report is first loaded.
This is because it loading and processing the data for all plots at once. To mitigate this,
large reports may show plots as grey boxes with a <em>&quot;Show Plot&quot;</em> button. Clicking this will
render the plot as normal and prevents the browser from trying to do everything at once.</p>
<p>By default this behaviour kicks in when a plot has 50 samples or more. This can be customised
by changing the <code>num_datasets_plot_limit</code> config option.</p>
<h3 id="flat--interactive-plots"><a href="#flat--interactive-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Flat / interactive plots</h3>
<p>Reports with many samples start to need a lot of data for plots. This results in inconvenient
report file sizes (can be 100s of megabytes) and worse, web browser crashes. To allow MultiQC
to scale to these sample numbers, most plot types have two plotting functions in the code base -
interactive (using HighCharts) and flat (rendered with MatPlotLib). Flat plots take up the
same disk space irrespective of sample number and do not consume excessive resources to display.</p>
<p>By default, MultiQC generates flat plots when there are 100 or more samples. This cutoff
can be changed by changing the <code>plots_flat_numseries</code> config option. This behaviour can also
be changed by running MultiQC with the <code>--flat</code> / <code>--interactive</code> command line options or by
setting the <code>plots_force_flat</code> / <code>plots_force_interactive</code> config options to <code>True</code>.</p>
<h3 id="tables--beeswarm-plots"><a href="#tables--beeswarm-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Tables / Beeswarm plots</h3>
<p>Report tables with thousands of samples (table rows) can quickly become impossible to use.
To avoid this, tables with large numbers of rows are instead plotted as a Beeswarm plot
(aka. a strip chart / jitter plot). These plots have fixed dimensions with any number
of samples. Hovering on a dot will highlight the same sample in other rows.</p>
<p>By default, MultiQC starts using beeswarm plots when a table has 500 rows or more. This
can be changed by setting the <code>max_table_rows</code> config option.</p></div><div class="docs_block" id="customisation.md"><h1 id="customising-reports"><a href="#customising-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Customising Reports</h1>
<p>MultiQC offers a few ways to customise reports to easily add your own
branding and some additional report-level information. These features
are primarily designed for core genomics facilities.</p>
<p>Note that much more extensive customisation of reports is possible using
<a href="http://multiqc.info/docs/#writing-new-templates">custom templates</a>.</p>
<h2 id="titles-and-introductory-text"><a href="#titles-and-introductory-text" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Titles and introductory text</h2>
<p>You can specify a custom title for the report using the <code>-t</code>/<code>--title</code>
command line option. The <code>-b</code>/<code>--comment</code> option can be used to add a
longer comment to the top of the report at run time.</p>
<p>You can also specify the title and comment, as well as
a subtitle and the introductory text in your config file:</p>
<pre><code class="language-yaml">title: "My Title"
subtitle: "A subtitle to go underneath in grey"
intro_text: "MultiQC reports summarise analysis results."
custom_message: "This is a comment about this report."</code></pre>
<p>Note that if <code>intro_text</code> is <code>None</code> the template will display the default
introduction sentence. Set this to <code>False</code> to hide this, or set it to a
string to use your own text.</p>
<h2 id="report-logo"><a href="#report-logo" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Report Logo</h2>
<p>To add your own custom logo to reports, you can add the following
three lines to your MultiQC configuration file:</p>
<pre><code class="language-yaml">custom_logo: '/abs/path/to/logo.png'
custom_logo_url: 'https://www.example.com'
custom_logo_title: 'Our Institute Name'</code></pre>
<p>Only <code>custom_logo</code> is needed. The URL will make the logo open up
a new web browser tab with your address and the title sets the mouse
hover title text.</p>
<h2 id="project-level-information"><a href="#project-level-information" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Project level information</h2>
<p>You can add custom information at the top of reports by adding key:value
pairs to the config option <code>report_header_info</code>. Note that if you have
a file called <code>multiqc_config.yaml</code> in the working directory, this will
automatically be parsed and added to the config. For example, if you have
the following saved:</p>
<pre><code class="language-yaml">report_header_info:
    - Contact E-mail: 'phil.ewels@scilifelab.se'
    - Application Type: 'RNA-seq'
    - Project Type: 'Application'
    - Sequencing Platform: 'HiSeq 2500 High Output V4'
    - Sequencing Setup: '2x125'</code></pre>
<p>Then this will be displayed at the top of reports:</p>
<p><img src="images/report_proj_info.png" alt="report project info" /></p>
<p>Note that you can also specify a path to a config file using <code>-c</code>.</p>
<h2 id="order-of-modules"><a href="#order-of-modules" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Order of modules</h2>
<p>By default, modules are included in the report as in the order specified in <code>config.module_order</code>.
Any modules found which aren't in this list are appended at the top of the report. To specify
certain modules that should always come at the top of the report, you can configure <code>config.top_modules</code>
in your MultiQC configuration file. For example, to always have the FastQC module at the top
of reports, add the following to your <code>~/.multiqc_config.yaml</code> file:</p>
<pre><code class="language-yaml">top_modules:
    - 'fastqc'</code></pre>
<h2 id="customising-tables"><a href="#customising-tables" class="header-link"><span class="glyphicon glyphicon-link"></span></a> Customising tables</h2>
<p>Report tables such as the General Statistics table can get quite wide. To help with this,
columns in the report can be hidden. Some MultiQC modules include columns which are hidden
by default, others may be uninteresting to some users.</p>
<p>To allow customisation of this behaviour, the defaults can be changed by adding to your
MultiQC config file. This is done with the <code>table_columns_visible</code> value. Open a MultiQC
report and click <em>Configure Columns</em> above a table. Make a note of the <em>Group</em> and <em>ID</em>
for the column that you'd like to alter. For example, to make the <code>% Duplicate Reads</code>
column from FastQC hidden by default, the <em>Group</em> is <code>FastQC</code> and the <em>ID</em> is
<code>percent_duplicates</code>. These are then added to the config as follows:</p>
<pre><code class="language-yaml">table_columns_visible:
    FastQC:
        percent_duplicates: False</code></pre>
<p>Note that you can set these to <code>True</code> to show columns that would otherwise be hidden
by default.</p>
<h2 id="number-formatting"><a href="#number-formatting" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Number formatting</h2>
<p>By default, the interactive HighCharts plots in MultiQC reports use spaces for thousand
separators and points for decimal places (<em>e.g.</em> <code>1 234 567.89</code>). Different countries
have different preferences for this, so you can customise the two using a couple of
configuration parameters - <code>decimalPoint_format</code> and <code>thousandsSep_format</code>.</p>
<p>For example, the following config would result in the following alternative
number formatting: <code>1234567,89</code>.</p>
<pre><code class="language-yaml">decimalPoint_format: ','
thousandsSep_format: ''</code></pre>
<p>This formatting currently only applies to the interactive charts. It may be extended
to apply elsewhere in the future (submit a new issue if you spot somewhere where you'd like it).</p>
<h2 id="troubleshooting"><a href="#troubleshooting" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Troubleshooting</h2>
<p>One tricky bit that caught me out whilst writing this is the different type casting
between Python, YAML and Jinja2 templates. This is especially true when using an
empty variable:</p>
<pre><code class="language-python"># Python
my_var = None</code></pre>
<pre><code class="language-yaml"># YAML
my_var: null</code></pre>
<pre><code class="language-python"># Jinja2
if myvar is none # Note - Lower case!</code></pre></div><div class="docs_block" id="troubleshooting.md"><h1 id="troubleshooting-1"><a href="#troubleshooting-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Troubleshooting</h1>
<p>Hopefully MultiQC will be easy to use and run without any hitches. If you have
any problems, please do get in touch with the developer
(<a href="http://phil.ewels.co.uk">Phil Ewels</a>) by e-mail or by
<a href="https://github.com/ewels/MultiQC/issues/new">submitting an issue</a> on github.
Before that, here are a few things previously encountered that may help...</p>
<h2 id="not-enough-samples-found"><a href="#not-enough-samples-found" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Not enough samples found</h2>
<p>In this scenario, MultiQC finds <em>some</em> logs for the bioinformatics tool
in question, but not all of your samples appear in the report. This is
the most common question I get regarding MultiQC operation.</p>
<p>Usually, this happens because sample names collide. This happens innocently
a lot - MultiQC overwrites previous results of the same name and you get
the last one seen in the report. You can see warnings about this by running
MultiQC in verbose mode with the <code>-v</code> flag, or looking at the generated log
file in <code>multiqc_data/multiqc.log</code>. If you are unsure about what log file
ended up in the report, look at <code>multiqc_data/multiqc_sources.txt</code> which
lists each source file used.</p>
<p>To solve this, try running MultiQC with the <code>-d</code> and <code>-s</code> flags.
The <a href="http://multiqc.info/docs/#clashing-sample-names">Clashing sample names</a>
section of the docs explains this in more detail.</p>
<h2 id="no-logs-found-for-a-tool"><a href="#no-logs-found-for-a-tool" class="header-link"><span class="glyphicon glyphicon-link"></span></a>No logs found for a tool</h2>
<p>In this case, you have run a bioinformatics tool and have some log files in
a directory. When you run MultiQC with that directory, it finds nothing
for the tool in question.</p>
<p>There are a couple of things you can check here:</p>
<ol>
<li>Is the tool definitely
<a href="https://github.com/ewels/MultiQC">supported by MultiQC</a>? If not, why
not <a href="https://github.com/ewels/MultiQC/issues/new">open an issue</a> to
request it!</li>
<li>Did your bioinformatics tool definitely run properly? I've spent quite
a bit of time debugging MultiQC modules only to realise that the output
files from the tool were empty or incomplete. If your data is missing,
take a look and the raw files and make sure that there's something to see!</li>
</ol>
<p>If everything looks fine, then MultiQC probably needs extending to support
your data. Tools have different versions, different parameters and different
output formats that can confuse the parsing code.
Please <a href="https://github.com/ewels/MultiQC/issues/new">open an issue</a> with
your log files and we can get it fixed.</p>
<h2 id="error-messages-about-mkl-trial-mode--licences"><a href="#error-messages-about-mkl-trial-mode--licences" class="header-link"><span class="glyphicon glyphicon-link"></span></a> Error messages about mkl trial mode / licences</h2>
<p>In this case you run MultiQC and get something like this:</p>
<pre><code>$ multiqc .

Vendor:  Continuum Analytics, Inc.
Package: mkl
Message: trial mode EXPIRED 2 days ago

    You cannot run mkl without a license any longer.
    A license can be purchased it at: http://continuum.io
    We are sorry for any inconveniences.

    SHUTTING DOWN PYTHON INTERPRETER</code></pre>
<p>The <code>mkl</code> library provides optimisations for <code>numpy</code>, a requirement of
<code>MatPlotLib</code>. Recent versions of Conda have a bundled version which should
come with a licence and remove the warning. See
<a href="https://docs.continuum.io/mkl-optimizations/index#dismissing-mkl-trial-warnings">this page</a>
for more info. If you already have Conda installed you can get the updated
version by running: </p>
<pre><code>conda remove mkl-rt
conda install -f mkl</code></pre>
<p>Another way around it is to uninstall <code>mkl</code>. It seems that <code>numpy</code> works
without it fine:</p>
<pre><code>$ conda remove --features mkl</code></pre>
<p>Problem solved! See more
<a href="http://stackoverflow.com/questions/25204021/anaconda-running-python-cannot-run-mkl-without-a-license">here</a> and
<a href="https://www.continuum.io/blog/developer-blog/anaconda-25-release-now-mkl-optimizations">here</a>.</p>
<p>If you're not using Conda, try installing MultiQC with that instead. You
can find instructions <a href="http://multiqc.info/docs/#installing-with-conda">here</a>.</p>
<h2 id="locale-error-messages"><a href="#locale-error-messages" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Locale Error Messages</h2>
<p>A more obscure problem is an error from the MatPlotLib python library
about locale settings. Some strings (such as <code>en_SE</code>) aren't allowed.
Running MultiQC gives the following error:</p>
<pre><code class="language-bash">$ multiqc --version</code></pre>
<pre><code class="language-python"># ..long traceback.. #
 File "/sw/comp/python/2.7.6_milou/lib/python2.7/locale.py", line 443, in _parse_localename
   raise ValueError, 'unknown locale: %s' % localename
ValueError: unknown locale: UTF-8</code></pre>
<p>You can fix this by changing the locale to something that will be
recognised. One way to do this is by adding these lines to your
<code>.bashrc</code> in your home directory (or <code>.bash_profile</code>):</p>
<pre><code class="language-bash">export LC_ALL=en_US.UTF-8
export LANG=en_US.UTF-8</code></pre></div></div><div class="docs_section">
<h1 class="section-header" id="multiqc-modules"><a href="#multiqc-modules" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC Modules</h1>
<div class="docs_block modules_block" id="pre-alignment-section"><h1 id="pre-alignment"><a href="#pre-alignment" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Pre-alignment</h1><h2 id="cutadapt"><a href="#cutadapt" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Cutadapt</h2><p class="mod_ext_link"><a href="https://code.google.com/p/cutadapt/" target="_blank">https://code.google.com/p/cutadapt/</a></p><p>The Cutadapt module parses results generated by
<a href="https://code.google.com/p/cutadapt/">Cutadapt</a>,
a tool to find and remove adapter sequences, primers, poly-A
tails and other types of unwanted sequence from your high-throughput
sequencing reads.</p>
<p>This module should be able to parse logs from a wide range of versions of Cutadapt.
It's been tested with log files from v1.2.1, 1.6 and 1.8. Note that you will need
to change the search pattern for very old log files (such as v.1.2) with the following
MultiQC config:</p>
<pre><code class="language-yaml">sp:
    cutadapt:
        contents: 'cutadapt version'</code></pre>
<p>See the <a href="http://multiqc.info/docs/#module-search-patterns">module search patterns</a>
section of the MultiQC documentation for more information.</p><h2 id="fastqc"><a href="#fastqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>FastQC</h2><p class="mod_ext_link"><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/" target="_blank">http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></p><p>The FastQC module parses results generated by
<a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>,
a  quality control tool for high throughput sequence data written
by Simon Andrews at the Babraham Institute.</p>
<p>FastQC generates a HTML report which is what most people use when
they run the program. However, it also helpfully generates a file
called <code>fastqc_data.txt</code> which is relatively easy to parse.</p>
<p>A typical run will produce the following files:</p>
<pre><code>mysample_fastqc.html
mysample_fastqc/
  Icons/
  Images/
  fastqc.fo
  fastqc_data.txt
  fastqc_report.html
  summary.txt</code></pre>
<p>Sometimes the directory is zipped, with just <code>mysample_fastqc.zip</code>.</p>
<p>The FastQC MultiQC module looks for files called <code>fastqc_data.txt</code>
or ending in <code>_fastqc.zip</code>. If the zip files are found, they are
read in memory and <code>fastqc_data.txt</code> parsed.</p>
<blockquote>
<p><strong>Note:</strong> The directory and zip file are often both present. To speed
up MultiQC execution, zip files will be skipped if the file name suggests
that they will share a sample name with data that has already been parsed.</p>
</blockquote>
<p>You can customise the patterns used for finding these files in your
MultiQC config using the following base - see the docs for more
details..</p>
<pre><code class="language-yaml">sp:
    fastqc:
        data:
            fn: 'fastqc_data.txt'
        zip:
            fn: '_fastqc.zip'</code></pre>
<blockquote>
<p><strong>Note:</strong> Sample names are discovered by parsing the line beginning
<code>Filename</code> in <code>fastqc_data.txt</code>, <em>not</em> based on the FastQC report names.</p>
</blockquote>
<h3 id="theoretical-gc-content"><a href="#theoretical-gc-content" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Theoretical GC Content</h3>
<p>It is possible to plot a dashed line showing the theoretical GC content for a
reference genome. MultiQC comes with genome and transcriptome guides for Human
and Mouse. You can use these in your reports by adding the following MultiQC
config keys (see <a href="http://multiqc.info/docs/#configuring-multiqc">Configuring MultiQC</a>):</p>
<pre><code class="language-yaml">fastqc_config:
    fastqc_theoretical_gc: 'hg38_genome'</code></pre>
<p>Only one theoretical distribution can be plotted. The following guides are available:
<code>hg38_genome</code>, <code>hg38_txome</code>, <code>mm10_genome</code>, <code>mm10_txome</code> (txome = transcriptome).</p>
<p>Alternatively, a custom theoretical guide can be used in reports. To do this,
create a file with <code>fastqc_theoretical_gc</code> in the filename and place it with your
analysis files. It should be tab delimited with the following format (column 1 = %GC,
column 2 = % of genome):</p>
<pre><code class="language-bash"># FastQC theoretical GC content curve: YOUR REFERENCE NAME
0   0.005311768
1   0.004108502
2   0.004060371
3   0.005066476
[...]</code></pre>
<p>You can generate these files using an R package called
<a href="https://github.com/mikelove/fastqcTheoreticalGC">fastqcTheoreticalGC</a>
written by <a href="https://github.com/mikelove">Mike Love</a>.
Please see the <a href="https://github.com/mikelove/fastqcTheoreticalGC">package readme</a>
for more details.</p>
<p>If you want to always use your custom file for MultiQC reports without having to
add it to the analysis directory, add the full file path to the same MultiQC config
variable described above:</p>
<pre><code class="language-yaml">fastqc_config:
    fastqc_theoretical_gc: '/path/to/your/custom_fastqc_theoretical_gc.txt'</code></pre><h2 id="fastq-screen"><a href="#fastq-screen" class="header-link"><span class="glyphicon glyphicon-link"></span></a>FastQ Screen</h2><p class="mod_ext_link"><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/" target="_blank">http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/</a></p><p>The FastQ Screen module parses results generated by
<a href="http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/">FastQ Screen</a>,
a tool that allows you to screen a library of sequences in FastQ format
against a set of sequence databases so you can see if the composition of
the library matches with what you expect.</p><h2 id="skewer"><a href="#skewer" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Skewer</h2><p class="mod_ext_link"><a href="https://github.com/relipmoc/skewer" target="_blank">https://github.com/relipmoc/skewer</a></p><p>The Skewer module parses results generated by
<a href="https://github.com/relipmoc/skewer">Skewer</a>,
an adapter trimming tool specially designed for processing
next-generation sequencing (NGS) paired-end sequences.</p><h2 id="trimmomatic"><a href="#trimmomatic" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Trimmomatic</h2><p class="mod_ext_link"><a href="http://www.usadellab.org/cms/?page=trimmomatic" target="_blank">http://www.usadellab.org/cms/?page=trimmomatic</a></p><p>The Trimmomatic module parses results generated by
<a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a>,
a flexible read trimming tool for Illumina NGS data.</p></div><div class="docs_block modules_block" id="aligners-section"><h1 id="aligners"><a href="#aligners" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Aligners</h1><h2 id="bismark"><a href="#bismark" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bismark</h2><p class="mod_ext_link"><a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/" target="_blank">http://www.bioinformatics.babraham.ac.uk/projects/bismark/</a></p><p>The Bismark module parses logs generated by
<a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/">Bismark</a>,
a tool to map bisulfite converted sequence reads and determine
cytosine methylation states.</p><h2 id="bowtie-1"><a href="#bowtie-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bowtie 1</h2><p class="mod_ext_link"><a href="http://bowtie-bio.sourceforge.net/" target="_blank">http://bowtie-bio.sourceforge.net/</a></p><p>The Bowtie 1 module parses results generated by
<a href="http://bowtie-bio.sourceforge.net/">Bowtie</a>,
an ultrafast, memory-efficient short read aligner.</p><h2 id="bowtie-2"><a href="#bowtie-2" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bowtie 2</h2><p class="mod_ext_link"><a href="http://bowtie-bio.sourceforge.net/bowtie2/" target="_blank">http://bowtie-bio.sourceforge.net/bowtie2/</a></p><p>The Bowtie 2 module parses results generated by
<a href="http://bowtie-bio.sourceforge.net/bowtie2/">Bowtie 2</a>,
an ultrafast and memory-efficient tool for aligning sequencing
reads to long reference sequences.</p>
<p>Please note that the Bowtie 2 logs are difficult to parse as they don't contain
much extra information (such as what the input data was). A typical log looks like this:</p>
<pre><code>314537 reads; of these:
  314537 (100.00%) were paired; of these:
    111016 (35.30%) aligned concordantly 0 times
    193300 (61.46%) aligned concordantly exactly 1 time
    10221 (3.25%) aligned concordantly &gt;1 times
    ----
    111016 pairs aligned concordantly 0 times; of these:
      11377 (10.25%) aligned discordantly 1 time
    ----
    99639 pairs aligned 0 times concordantly or discordantly; of these:
      199278 mates make up the pairs; of these:
        112779 (56.59%) aligned 0 times
        85802 (43.06%) aligned exactly 1 time
        697 (0.35%) aligned &gt;1 times
82.07% overall alignment rate</code></pre>
<p>Bowtie 2 logs are from <code>STDERR</code> - some pipelines (such as <a href="http://clusterflow.io">Cluster Flow</a>)
print the Bowtie 2 command before this, so MultiQC looks to see if this can be recognised in the same
file. If not, it takes the filename as the sample name.</p>
<p>Bowtie 2 is used by other tools too, so if your log file contains the word <code>bisulfite</code>, MultiQC will
assume that this is actually Bismark and ignore the Bowtie 2 logs.</p><h2 id="hicup"><a href="#hicup" class="header-link"><span class="glyphicon glyphicon-link"></span></a>HiCUP</h2><p class="mod_ext_link"><a href="http://www.bioinformatics.babraham.ac.uk/projects/hicup/" target="_blank">http://www.bioinformatics.babraham.ac.uk/projects/hicup/</a></p><p>The HiCUP module parses results generated by
<a href="http://www.bioinformatics.babraham.ac.uk/projects/hicup/">HiCUP</a>,
(Hi-C User Pipeline), a tool for mapping and performing quality
control on Hi-C data.</p><h2 id="kallisto"><a href="#kallisto" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Kallisto</h2><p class="mod_ext_link"><a href="http://pachterlab.github.io/kallisto/" target="_blank">http://pachterlab.github.io/kallisto/</a></p><p>The Kallisto module parses logs generated by
<a href="http://pachterlab.github.io/kallisto/">Kallisto</a>,
a program for quantifying abundances of transcripts from RNA-Seq data, or more generally
of target sequences using high-throughput sequencing reads.</p><h2 id="star"><a href="#star" class="header-link"><span class="glyphicon glyphicon-link"></span></a>STAR</h2><p class="mod_ext_link"><a href="https://github.com/alexdobin/STAR" target="_blank">https://github.com/alexdobin/STAR</a></p><p><a href="https://github.com/alexdobin/STAR">STAR</a> is an ultrafast universal RNA-seq aligner.</p>
<p>This MultiQC module parses summary statistics from the <code>Log.final.out</code> log files.
Sample names are taken either from the filename prefix (<code>sampleNameLog.final.out</code>)
when set with <code>--outFileNamePrefix</code> in STAR. If there is no filename prefix,
the sample name is set as the name of the directory containing the file.</p>
<p>In addition to this summary log file, the module parses <code>ReadsPerGene.out.tab</code>
files generated with <code>--quantMode GeneCounts</code>, if found. </p><h2 id="salmon"><a href="#salmon" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Salmon</h2><p class="mod_ext_link"><a href="http://combine-lab.github.io/salmon/" target="_blank">http://combine-lab.github.io/salmon/</a></p><p>The Salmon module parses results generated by
<a href="http://combine-lab.github.io/salmon/">Salmon</a>,
a tool for quantifying the expression of transcripts using RNA-seq data.</p><h2 id="tophat"><a href="#tophat" class="header-link"><span class="glyphicon glyphicon-link"></span></a>TopHat</h2><p class="mod_ext_link"><a href="https://ccb.jhu.edu/software/tophat/" target="_blank">https://ccb.jhu.edu/software/tophat/</a></p><p>The TopHat module parses results generated by
<a href="https://ccb.jhu.edu/software/tophat/">TopHat</a>,
a fast splice junction mapper for RNA-Seq reads that aligns
RNA-Seq reads to mammalian-sized genomes.</p></div><div class="docs_block modules_block" id="post-alignment-section"><h1 id="post-alignment"><a href="#post-alignment" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Post-alignment</h1><h2 id="bamtools"><a href="#bamtools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bamtools</h2><p class="mod_ext_link"><a href="https://github.com/pezmaster31/bamtools" target="_blank">https://github.com/pezmaster31/bamtools</a></p><p>The Bamtools module parses <code>bamtools stats</code> logs generated by
<a href="https://github.com/pezmaster31/bamtools">Bamtools</a>,
a programmer's API and an end-user's toolkit for handling BAM files.</p>
<p>Supported commands: <code>stats</code></p><h2 id="bcftools"><a href="#bcftools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bcftools</h2><p class="mod_ext_link"><a href="https://samtools.github.io/bcftools/" target="_blank">https://samtools.github.io/bcftools/</a></p><p>The Bcftools module parses results generated by
<a href="https://samtools.github.io/bcftools/">Bcftools</a>,
a suite of programs for interacting with variant call data.</p>
<p>Supported commands: <code>stats</code></p><h2 id="featurecounts"><a href="#featurecounts" class="header-link"><span class="glyphicon glyphicon-link"></span></a>featureCounts</h2><p class="mod_ext_link"><a href="http://bioinf.wehi.edu.au/featureCounts/" target="_blank">http://bioinf.wehi.edu.au/featureCounts/</a></p><p>The featureCounts module parses results generated by
<a href="http://bioinf.wehi.edu.au/featureCounts/">featureCounts</a>,
a highly efficient general-purpose read summarization
program that counts mapped reads for genomic features such as genes,
exons, promoter, gene bodies, genomic bins and chromosomal locations.</p><h2 id="gatk"><a href="#gatk" class="header-link"><span class="glyphicon glyphicon-link"></span></a>GATK</h2><p class="mod_ext_link"><a href="https://www.broadinstitute.org/gatk/" target="_blank">https://www.broadinstitute.org/gatk/</a></p><p>Developed by the <a href="http://www.broadinstitute.org/dsde">Data Science and Data Engineering</a>
group at the <a href="http://www.broadinstitute.org/">Broad Institute</a>, the GATK toolkit offers
a wide variety of tools with a primary focus on variant discovery and genotyping.
Its powerful processing engine and high-performance computing features make it capable
of taking on projects of any size.</p>
<p>Supported tools: <code>VariantEval</code></p>
<h3 id="varianteval"><a href="#varianteval" class="header-link"><span class="glyphicon glyphicon-link"></span></a>VariantEval</h3>
<p><a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_varianteval_VariantEval.php">VariantEval</a>
is a general-purpose tool for variant evaluation. It gives information about percentage of
variants in dbSNP, genotype concordance, Ti/Tv ratios and a lot more.</p><h2 id="htseq"><a href="#htseq" class="header-link"><span class="glyphicon glyphicon-link"></span></a>HTSeq</h2><p class="mod_ext_link"><a href="http://www-huber.embl.de/HTSeq/doc/overview.html" target="_blank">http://www-huber.embl.de/HTSeq/doc/overview.html</a></p><p>HTSeq is a general purpose Python package that provides infrastructure to
process data from high-throughput sequencing assays. <code>htseq-count</code> is a tool
that is part of the main HTSeq package - it takes a file with aligned sequencing
reads, plus a list of genomic features and counts how many reads map to each feature.</p><h2 id="methylqa"><a href="#methylqa" class="header-link"><span class="glyphicon glyphicon-link"></span></a>methylQA</h2><p class="mod_ext_link"><a href="http://methylqa.sourceforge.net/" target="_blank">http://methylqa.sourceforge.net/</a></p><p>The methylQA module parses results generated by
<a href="http://methylqa.sourceforge.net/">methylQA</a>,
a methylation sequencing data quality assessment tool.</p><h2 id="peddy"><a href="#peddy" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Peddy</h2><p class="mod_ext_link"><a href="https://github.com/brentp/peddy" target="_blank">https://github.com/brentp/peddy</a></p><p>Peddy compares familial-relationships and sexes as reported in a
PED file with those inferred from a VCF.</p>
<p>It samples the VCF at about 25000 sites (plus chrX) to accurately
estimate relatedness, IBS0, heterozygosity, sex and ancestry.
It uses 2504 thousand genome samples as backgrounds to calibrate
the relatedness calculation and to make ancestry predictions.</p>
<p>It does this very quickly by sampling, by using C for computationally
intensive parts, and by parallelization.</p><h2 id="picard"><a href="#picard" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Picard</h2><p class="mod_ext_link"><a href="http://broadinstitute.github.io/picard/" target="_blank">http://broadinstitute.github.io/picard/</a></p><p>The Picard module parses results generated by
<a href="http://broadinstitute.github.io/picard/">Picard</a>,
a set of Java command line tools for manipulating high-throughput
sequencing data.</p>
<p>Supported commands:</p>
<ul>
<li><code>MarkDuplicates</code></li>
<li><code>InsertSizeMetrics</code></li>
<li><code>GcBiasMetrics</code></li>
<li><code>HsMetrics</code></li>
<li><code>OxoGMetrics</code></li>
<li><code>BaseDistributionByCycl</code></li>
<li><code>RnaSeqMetrics</code></li>
<li><code>AlignmentSummaryMetrics</code></li>
</ul>
<h4>HsMetrics</h4>
<p>It's possible to customise the HsMetrics <em>&quot;Target Bases 30X&quot;</em> coverage that's
shown in the general statistics table. This must correspond to field names in the
HsMetrics report, such as <code>PCT_TARGET_BASES_2X</code>. Any numbers not found in the
reports will be ignored. The coverage levels available are
<a href="https://github.com/ewels/MultiQC_TestData/blob/master/data/modules/picard/HsMetrics/alignment-hs_metrics.txt">typically</a>
2, 10, 20, 30, 40, 50 and 100X.</p>
<p>To customise this, add something like the following to your MultiQC config:</p>
<pre><code class="language-yaml">picard_config:
    general_stats_target_coverage:
        - 2
        - 10</code></pre><h2 id="preseq"><a href="#preseq" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Preseq</h2><p class="mod_ext_link"><a href="http://smithlabresearch.org/software/preseq/" target="_blank">http://smithlabresearch.org/software/preseq/</a></p><p>The Preseq module parses results generated by
<a href="http://smithlabresearch.org/software/preseq/">Preseq</a>,
a tool that estimates the complexity of a library, showing how many additional
unique reads are sequenced for increasing total read count.</p><h2 id="prokka"><a href="#prokka" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Prokka</h2><p class="mod_ext_link"><a href="http://www.vicbioinformatics.com/software.prokka.shtml" target="_blank">http://www.vicbioinformatics.com/software.prokka.shtml</a></p><p>The Prokka module analyses summary results from the
<a href="http://www.vicbioinformatics.com/software.prokka.shtml">Prokka</a> annotation
pipeline for prokaryotic genomes.<br />
The Prokka module accepts two configuration options:</p>
<ul>
<li><code>prokka_table</code>: default <code>False</code>. Show a table in the report.</li>
<li><code>prokka_barplot</code>: default <code>True</code>. Show a barplot in the report.</li>
</ul><h2 id="qualimap"><a href="#qualimap" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Qualimap</h2><p class="mod_ext_link"><a href="http://qualimap.bioinfo.cipf.es/" target="_blank">http://qualimap.bioinfo.cipf.es/</a></p><p>The Qualimap module parses results generated by
<a href="http://qualimap.bioinfo.cipf.es/">Qualimap</a>,
a platform-independent application to facilitate the quality
control of alignment sequencing data and its derivatives like feature
counts.</p>
<p>Supported commands: <code>BamQC</code>, <code>RNASeq</code></p>
<p>It is possible to customise which coverage thresholds are shown from BamQC in
the General Statistics table <em>(default: 1, 5, 10, 30, 50)</em> and which
of these are hidden when the report loads <em>(default: all except 30X)</em>.</p>
<p>To do this, add something like the following to your MultiQC config file:</p>
<pre><code class="language-yaml">qualimap_config:
    general_stats_coverage:
        - 10
        - 20
        - 40
        - 200
        - 30000
    general_stats_coverage_hidden:
        - 10
        - 20
        - 200</code></pre><h2 id="quast"><a href="#quast" class="header-link"><span class="glyphicon glyphicon-link"></span></a>QUAST</h2><p class="mod_ext_link"><a href="http://quast.bioinf.spbau.ru/" target="_blank">http://quast.bioinf.spbau.ru/</a></p><p><a href="http://quast.bioinf.spbau.ru/">QUAST</a> evaluates genome assemblies by
computing various metrics, including</p>
<ul>
<li>N50, length for which the collection of all contigs of that length or
longer covers at least 50% of assembly length</li>
<li>NG50, where length of the reference genome is being covered</li>
<li>NA50 and NGA50, where aligned blocks instead of contigs are taken</li>
<li>Misassemblies, misassembled and unaligned contigs or contigs bases</li>
<li>Genes and operons covered</li>
</ul>
<p>The QUAST MultiQC module parses the <code>report.tsv</code> files generated by
QUAST and adds key metrics to the report General Statistics table.
All statistics for all samples are saved to <code>multiqc_data/multiqc_quast.txt</code>.</p><h2 id="rseqc"><a href="#rseqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>RSeQC</h2><p class="mod_ext_link"><a href="http://rseqc.sourceforge.net/" target="_blank">http://rseqc.sourceforge.net/</a></p><p>The RSeQC module parses results generated by
<a href="http://rseqc.sourceforge.net/">RSeQC</a>,
a package that provides a number of useful modules that can
comprehensively evaluate high throughput RNA-seq data.</p>
<p>Supported scripts:</p>
<ul>
<li><code>bam_stat</code></li>
<li><code>gene_body_coverage</code></li>
<li><code>infer_experiment</code></li>
<li><code>inner_distance</code></li>
<li><code>junction_annotation</code></li>
<li><code>junction_saturation</code></li>
<li><code>read_distribution</code></li>
<li><code>read_duplication</code></li>
<li><code>read_gc</code></li>
</ul><h2 id="samblaster"><a href="#samblaster" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Samblaster</h2><p class="mod_ext_link"><a href="https://github.com/GregoryFaust/samblaster" target="_blank">https://github.com/GregoryFaust/samblaster</a></p><p>The Samblaster module parses results generated by
<a href="https://github.com/GregoryFaust/samblaster">Samblaster</a>,
a tool to mark duplicates and extract discordant and split
reads from sam files.</p><h2 id="samtools"><a href="#samtools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Samtools</h2><p class="mod_ext_link"><a href="http://www.htslib.org" target="_blank">http://www.htslib.org</a></p><p>The Samtools module parses results generated by
<a href="http://www.htslib.org">Samtools</a>,
a suite of programs for interacting with high-throughput
sequencing data.</p>
<p>Supported commands:</p>
<ul>
<li><code>stats</code></li>
<li><code>flagstats</code></li>
<li><code>idxstats</code></li>
<li><code>rmdup</code></li>
</ul>
<h3 id="idxstats"><a href="#idxstats" class="header-link"><span class="glyphicon glyphicon-link"></span></a>idxstats</h3>
<p>The <code>samtools idxstats</code> prints its results to standard
out (no consistent file name) and has no header lines
(no way to recognise from content of file). As such, <code>idxstats</code>
result files must have the string <code>idxstat</code> somewhere in the filename.</p>
<p>There are a few MultiQC config options that you can add to
customise how the idxstats module works. A typical configuration
could look as follows:</p>
<pre><code class="language-yaml"># Always include these chromosomes in the plot
samtools_idxstats_always:
    - X
    - Y

# Never include these chromosomes in the plot
samtools_idxstats_ignore:
    - MT

# Threshold where chromosomes are ignored in the plot.
# Should be a fraction, default is 0.001 (0.1% of total)
samtools_idxstats_fraction_cutoff: 0.001

# Name of the X and Y chromosomes.
# If not specified, MultiQC will search for any chromosome
# names that look like x, y, chrx or chry (case insensitive search)
samtools_idxstats_xchr: myXchr
samtools_idxstats_ychr: myYchr</code></pre><h2 id="slamdunk"><a href="#slamdunk" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Slamdunk</h2><p class="mod_ext_link"><a href="http://t-neumann.github.io/slamdunk/" target="_blank">http://t-neumann.github.io/slamdunk/</a></p><p><a href="http://t-neumann.github.io/slamdunk/">Slamdunk</a> is a tool to analyze data from the SLAM-Seq sequencing protocol.</p>
<p>This module should be able to parse logs from v0.2.2-dev onwards.</p><h2 id="snpeff"><a href="#snpeff" class="header-link"><span class="glyphicon glyphicon-link"></span></a>SnpEff</h2><p class="mod_ext_link"><a href="http://snpeff.sourceforge.net/" target="_blank">http://snpeff.sourceforge.net/</a></p><p>The SnpEff module parses results generated by
<a href="http://snpeff.sourceforge.net/">SnpEff</a>,
a genetic variant annotation and effect prediction toolbox.
It annotates and predicts the effects of variants on genes (such as amino
acid changes).</p></div></div><div class="docs_section">
<h1 class="section-header" id="custom-content"><a href="#custom-content" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom Content</h1>
<div class="docs_block" id="custom_content.md"><blockquote>
<p><strong>WARNING</strong> - This feature is new and is very much in a beta status.
It is expected to be further developed in future releases, which may break backwards
compatibility. There are also probably quite a few bugs. Use at your own risk!
Please report bugs or missing functionality as a
<a href="https://github.com/ewels/MultiQC/issues/new">new GitHub issue</a>.</p>
</blockquote>
<h1 id="introduction"><a href="#introduction" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Introduction</h1>
<p>Bioinformatics projects often include non-standardised analyses, with results from custom
scripts or in-house packages. It can be frustrating to have a MultiQC report describing
results from 90% of your pipeline but missing the final key plot. To help with this,
MultiQC has a special <em>&quot;custom content&quot;</em> module.</p>
<p>Custom content parsing is a little more restricted than standard modules. Specifically:</p>
<ul>
<li>Only one plot per section is possible</li>
<li>Plot customisation is more limited</li>
</ul>
<p>All plot types can be generated using custom content - see the
<a href="https://github.com/ewels/MultiQC_TestData/tree/master/data/custom_content">test files</a>
for examples of how data should be structured.</p>
<h1 id="configuration"><a href="#configuration" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Configuration</h1>
<p>Data should typically be submitted alongside some configuration, to specify how
MultiQC should parse and display the data. All of these configuration parameters
are optional, and MultiQC will do its best to guess sensible defaults if they are
not specified.</p>
<p>All possible configuration keys and their default values are shown below:</p>
<pre><code class="language-yaml">id: null                # Unique ID for report section.
section_anchor: &lt;id&gt;    # Used in report section #soft-links
section_name: &lt;id&gt;      # Nice name used for the report section header
section_href: null      # External URL for the data, to find more information
description: null       # Introductory text to be printed under the section header
file_format: null       # File format of the data (typically csv / tsv - see below for more information)
plot_type: null         # The plot type to visualise the data with.
                        # - Possible options: generalstats | table | bargraph | linegraph | scatter | heatmap | beeswarm
pconfig: {}             # Configuration for the plot. See http://multiqc.info/docs/#plotting-functions</code></pre>
<p>Note that any <em>custom content</em> data found with the same section <code>id</code> will be merged
into the same report section / plot. The other section configuration keys are merged
for each file, with identical keys overwriting what was previously parsed.</p>
<p>This approach means that it's possible to have a single file containing data for multiple
samples, but it's also possible to have one file per sample and still have all of them
summarised.</p>
<p>If you're using <code>plot_type: 'generalstats'</code> then a report section will not be created and
most of the configuration keys above are ignored.</p>
<p>Data types <code>generalstats</code> and <code>beeswarm</code> are <em>only</em> possible by setting the above
configuration keys (these can't be guessed by data format).</p>
<h1 id="data-formats"><a href="#data-formats" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Data formats</h1>
<p>MultiQC can parse custom data from a few different sources, in a number of different
formats. Which one you use depends on how the data is being produced.</p>
<p>A quick summary of which approach to use looks something like this:</p>
<ul>
<li>Additional data when already using custom MultiQC config files
<ul>
<li><em>Data as part of MultiQC config</em></li>
</ul></li>
<li>Data specifically for MultiQC from a custom script
<ul>
<li><em>MultiQC-specific data file</em></li>
</ul></li>
<li>Data from a custom script which is also used by other processes
<ul>
<li><em>Separate configuration and data files</em></li>
<li>Add <code>_mqc.txt</code> to filename and hope that MultiQC guesses correctly</li>
</ul></li>
<li>Anything more complicated, or data from a released tool
<ul>
<li>Write a proper MultiQC module instead.</li>
</ul></li>
</ul>
<p>For more complete examples of the data formats understood by MultiQC, please see the
<a href="https://github.com/ewels/MultiQC_TestData/tree/master/data/custom_content"><code>data/custom_content</code></a>
directory in the <a href="https://github.com/ewels/MultiQC_TestData">MultiQC_TestData</a>
GitHub repository.</p>
<h2 id="data-from-a-released-tool"><a href="#data-from-a-released-tool" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Data from a released tool</h2>
<p>If your data comes from a released bioinformatics tool, you shouldn't be using this
feature of MultiQC! Sure, you can probably get it to work, but it's better if a
fully-fledged core MultiQC module is written instead. That way, other users of MultiQC
can also benefit from results parsing.</p>
<p>Note that proper MultiQC modules are more robust and powerful than this custom-content
feature. You can also <a href="http://multiqc.info/docs/#writing-new-modules">write modules</a>
in <a href="http://multiqc.info/docs/#multiqc-plugins">MultiQC plugins</a> if they're not suitable for
general release.</p>
<h2 id="data-as-part-of-multiqc-config"><a href="#data-as-part-of-multiqc-config" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Data as part of MultiQC config</h2>
<p>If you are already using a MultiQC config file to add data to your report (for example,
<a href="http://multiqc.info/docs/#customising-reports">titles / introductory text</a>), you can
give data within this file too. This can be in any MultiQC config file (for example,
passed on the command line with <code>-c my_yaml_file.yaml</code>). This is useful as you can
keep everything contained within a single file (including stuff unrelated to this
specific <em>custom content</em> feature of MultiQC).</p>
<p>If you're not using this file for other MultiQC configuration, you're probably
better off using a stand-alone YAML file (see <a href="#multiqc-specific-data-file">section below</a>).</p>
<p>To be understood by MultiQC, the <code>custom_data</code> key must be found. This must contain
a section with a unique id, specific to your new report section. This in turn
must contain a section called <code>data</code>. Other configuration keys can be held alongside this.
For example:</p>
<pre><code class="language-yaml"># Other MultiQC config stuff here
custom_data:
    my_data_type:
        id: 'mqc_config_file_section'
        section_name: 'My Custom Section'
        description: 'This data comes from a single multiqc_config.yaml file'
        plot_type: 'bargraph'
        pconfig:
            id: 'barplot_config_only'
            title: 'MultiQC Config Data Plot'
            ylab: 'Number of things'
        data:
            sample_a:
                first_thing: 12
                second_thing: 14
            sample_b:
                first_thing: 8
                second_thing: 6
            sample_c:
                first_thing: 11
                second_thing: 5
            sample_d:
                first_thing: 12
                second_thing: 9</code></pre>
<p>Or to add data to the General Statistics table:</p>
<pre><code class="language-yaml">custom_data:
    my_genstats:
        plot_type: 'generalstats'
        pconfig:
            - col_1:
                max: 100
                min: 0
                scale: 'RdYlGn'
                format: '{:.1f}%'
            - col_2:
                min: 0
        data:
            sample_a:
                col_1: 14.32
                col_2: 1.2
            sample_b:
                col_1: 84.84
                col_2: 1.9</code></pre>
<blockquote>
<p><strong>Note:</strong> Use a <strong>list</strong> of headers in <code>pconfig</code> (keys prepended with <code>-</code>) to specify the order
of columns in the table.</p>
</blockquote>
<p>See the <a href="http://multiqc.info/docs/#step-4-adding-to-the-general-statistics-table">general statistics docs</a>
for more information about configuring data for the General Statistics table.</p>
<h2 id="multiqc-specific-data-file"><a href="#multiqc-specific-data-file" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC-specific data file</h2>
<p>If you can choose exactly how your data output looks, then the easiest way to parse it
is to use a MultiQC-specific format. If the filename ends in <code>*_mqc.(yaml|json|txt|csv|out)</code>
then it will be found by any standard MultiQC installation with no additional customisation
required (v0.9 onwards).</p>
<p>These files contain configuration information specifying how the data should be parsed, along
side the data. If using YAML, this looks just the same as if in a MultiQC config file (see above),
but without having to be within a <code>custom_data</code> section:</p>
<pre><code class="language-yaml">id: 'my_pca_section'
section_name: 'PCA Analysis'
description: 'This plot shows the first two components from a principal component analysis.'
plot_type: 'scatter'
pconfig:
    id: 'pca_scatter_plot'
    title: 'PCA Plot'
    xlab: 'PC1'
    ylab: 'PC2'
data:
    sample_1: {x: 12, y: 14}
    sample_2: {x: 8, y: 6 }
    sample_3: {x: 5, y: 11}
    sample_4: {x: 9, y: 12}</code></pre>
<p>The file format can also be JSON:</p>
<pre><code class="language-json">{
    "id": "custom_data_lineplot",
    "section_name": "Custom JSON File",
    "description": "This plot is a self-contained JSON file.",
    "plot_type": "linegraph",
    "pconfig": {
        "id": "custom_data_linegraph",
        "title": "Output from my JSON file",
        "ylab": "Number of things",
        "xDecimals": false
    },
    "data": {
        "sample_1": { "1": 12, "2": 14, "3": 10, "4": 7, "5": 16 },
        "sample_2": { "1": 9, "2": 11, "3": 15, "4": 18, "5": 21 }
    }
}</code></pre>
<p>If you want the data to be easy to use with other tools, you can also use
comma-separated or tab-separated file. To customise plot output, include commented
header lines with plot configuration in YAML format:</p>
<pre><code class="language-bash"># title: 'Output from my script'
# description: 'This output is described in the file header. Any MultiQC installation will understand it without prior configuration.'
# section: 'Custom Data File'
# format: 'tsv'
# plot_type: 'bargraph'
# pconfig:
#    id: 'custom_bargraph_w_header'
#    ylab: 'Number of things'
Category_1    374
Category_2    229
Category_3    39
Category_4    253</code></pre>
<p>If no configuration is given, MultiQC will do its best to guess how to visualise your data appropriately.
To see examples of typical file structures which are understood, see the
<a href="https://github.com/ewels/MultiQC_TestData/tree/master/data/custom_content/no_config">test data</a>
used to develop this code. Something will be probably be shown, but it may produce unexpected results.</p>
<h2 id="separate-configuration-and-data-files"><a href="#separate-configuration-and-data-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Separate configuration and data files</h2>
<p>It's not always possible or desirable to include MultiQC configuration within a data file.
If this is the case, you can add to the MultiQC configuration to specify how input files
should be parsed.</p>
<p>As described in the above <a href="#data-as-part-of-multiqc-config"><em>Data as part of MultiQC config</em></a> section,
this configuration should be held within a section called <code>custom_data</code> with a section-specific id.
The only difference is that no <code>data</code> subsection should be present and the <code>sp</code> key <em>must</em> be present.</p>
<p>The <code>sp</code> key tells MultiQC how to find files. This should have <code>fn</code> and/or <code>contents</code> under it, as
described in the <a href="http://multiqc.info/docs/#module-search-patterns">Module search patterns</a> section
of the docs.</p>
<p>For example:</p>
<pre><code class="language-yaml"># Other MultiQC config stuff here
custom_data:
    example_files:
        sp:
            fn: 'example_files_*'
        file_format: 'tsv'
        section_name: 'Coverage Decay'
        description: 'This plot comes from files acommpanied by a mutliqc_config.yaml file for configuration'
        plot_type: 'linegraph'
        pconfig:
            id: 'example_coverage_lineplot'
            title: 'Coverage Decay'
            ylab: 'X Coverage'
            ymax: 100
            ymin: 0</code></pre>
<p>A data file within the MultiQC search directories could then simply look like this:</p>
<p><code>example_files_Sample_1.txt</code>:</p>
<pre><code>0   98.22076066
1   97.96764159
2   97.78227175
3   97.61262195
[...]</code></pre>
<p>As mentioned above - if no configuration is given, MultiQC will do its best to guess how to visualise
your data appropriately. To see examples of typical file structures which are understood, see the
<a href="https://github.com/ewels/MultiQC_TestData/tree/master/data/custom_content/no_config">test data</a>
used to develop this code.</p></div></div><div class="docs_section">
<h1 class="section-header" id="coding-with-multiqc"><a href="#coding-with-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Coding with MultiQC</h1>
<div class="docs_block" id="templates.md"><h1 id="writing-new-templates"><a href="#writing-new-templates" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Writing New Templates</h1>
<p>MultiQC is built around a templating system that uses the
<a href="http://jinja.pocoo.org/">Jinja</a> python package. This makes it very
easy to create new report templates that fit your needs.</p>
<h2 id="core-or-plugin"><a href="#core-or-plugin" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Core or plugin?</h2>
<p>If your template could be of use to others, it would be great if you
could add it to the main MultiQC package. You can do this by creating a
fork of the <a href="https://github.com/ewels/MultiQC">MultiQC GitHub repository</a>,
adding your template and then creating a pull request to merge your changes
back to the main repository.</p>
<p>If it's very specific template, you can create a new Python package which
acts as a plugin. For more information about this, see the
<a href="http://multiqc.info/docs/#multiqc-plugins">plugins documentation</a>.</p>
<h2 id="creating-a-template-skeleton"><a href="#creating-a-template-skeleton" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Creating a template skeleton</h2>
<p>For a new template to be recognised by MultiQC, it must be a python submodule
directory with a <code>__init__.py</code> file. This must be referenced in the <code>setup.py</code>
installation script as an
<a href="https://pythonhosted.org/setuptools/setuptools.html#dynamic-discovery-of-services-and-plugins">entry point</a>.</p>
<p>You can see the bundled templates defined in this way:</p>
<pre><code class="language-python">entry_points = {
    'multiqc.templates.v1': [
        'default = multiqc.templates.default',
        'default_dev = multiqc.templates.default_dev',
        'simple = multiqc.templates.simple',
        'geo = multiqc.templates.geo',
    ]
}</code></pre>
<p>Note that these entry points can point to any Python modules, so if you're
writing a plugin module you can specify your module name instead. Just make
sure that <code>multiqc.templates.v1</code> is the same.</p>
<p>Once you've added the entry point, remember to install the package again:</p>
<pre><code>python setup.py develop</code></pre>
<p>Using <code>develop</code> tells setuptools to softlink the plugin files instead of
copying, so changes made whilst editing files will be reflected when you
run MultiQC.</p>
<p>The <code>__init__.py</code> files must define two variables - the path to the template
directory and the main jinja template file:</p>
<pre><code class="language-python">template_dir = os.path.dirname(__file__)
base_fn = 'base.html'</code></pre>
<h2 id="child-templates"><a href="#child-templates" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Child templates</h2>
<p>The default MultiQC template contains a <em>lot</em> of code. Importantly, it includes
1448 lines of custom JavaScript (at time of writing) which powers the plotting
and dynamic functions in the report. You probably don't want to rewrite all of
this for your template, so to make your life easier you can create a
<em>child template</em>.</p>
<p>To do this, add an extra variable to your template's <code>__init__.py</code>:</p>
<pre><code class="language-python">template_parent = 'default'</code></pre>
<p>This tells MultiQC to use the template files from the <code>default</code> template unless
a file with the same name is found in your child template. For instance, if
you just want to add your own logo in the header of the reports, you can create
your own <code>header.html</code> which will overwrite the default header.</p>
<p>Files within the default template have comments at the top explaining what
part of the report they generate.</p>
<h2 id="extra-init-variables"><a href="#extra-init-variables" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Extra init variables</h2>
<p>There are a few extra variables that can be added to the <code>__init__.py</code> file
to change how the report is generated.</p>
<p>Setting <code>output_dir</code> instructs MultiQC to put the report and it's contents
into a subdirectory. Set the string to your desired name. Note that this will
be prefixed if <code>-p</code>/<code>--prefix</code> is set at run time.</p>
<p>Secondly, you can copy additional files with your report when it is generated.
This is usually used to copy required images or scripts with the report. These
should be a list of file or directory paths, relative to the <code>__init__.py</code> file.
Directory contents will be copied recursively.</p>
<p>You can also override config options in the template. For example, setting
the value of <code>config.plots_force_flat</code> can force the report to only have
static image plots.</p>
<pre><code class="language-python">from multiqc.utils import config

output_subdir = 'multiqc_report'
copy_files = ['assets']
config.plots_force_flat = True</code></pre>
<h2 id="jinja-template-variables"><a href="#jinja-template-variables" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Jinja template variables</h2>
<p>There are a number of variables that you can use within your Jinja template.
Two namespaces are available - <code>report</code> and <code>config</code>. You can print these
using the Jinja curly brace syntax, <em>eg.</em> <code>{{ config.version }}</code>. See the
<a href="http://jinja.pocoo.org/docs/dev/templates/">Jinja2 documentation</a> for more
information.</p>
<p>The default MultiQC template includes dependencies in the HTML so that the
report is standalone. If you would like to do the same, use the <code>include_file</code>
function. For example:</p>
<pre><code class="language-html">&lt;script&gt;{{ include_file('js/jquery.min.js') }}&lt;/script&gt;
&lt;img src="data:image/png;base64,{{ include_file('img/logo.png', b64=True) }}"&gt;</code></pre>
<h2 id="appendices"><a href="#appendices" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Appendices</h2>
<h3 id="custom-plotting-functions"><a href="#custom-plotting-functions" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom plotting functions</h3>
<p>If you don't like the default plotting functions built into MultiQC, you
can write your own! If you create a callable variable in a template called
either <code>bargraph</code> or <code>linegraph</code>, MultiQC will use that instead. For example:</p>
<pre><code class="language-python">def custom_linegraph(plotdata, pconfig):
    return '&lt;h1&gt;Awesome line graph here&lt;/h1&gt;'
linegraph = custom_linegraph

def custom_bargraph(plotdata, plotseries, pconfig):
    return '&lt;h1&gt;Awesome bar graph here&lt;/h1&gt;'
bargraph = custom_bargraph</code></pre>
<p>These particular examples don't do very much, but hopefully you get the idea.
Note that you have to set the variable <code>linegraph</code> or <code>bargraph</code> to your function.</p></div><div class="docs_block" id="modules.md"><h1 id="writing-new-modules"><a href="#writing-new-modules" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Writing New Modules</h1>
<h2 id="introduction-1"><a href="#introduction-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Introduction</h2>
<p>Writing a new module can at first seem a daunting task. However, MultiQC
has been written <em>(and refactored)</em> to provide a lot of functionality
as common functions.</p>
<p>Provided that you are familiar with writing Python and you have a read
through the guide below, you should be on your way in no time!</p>
<p>If you have any problems, feel free to contact the author - details
here: <a href="https://github.com/ewels">@ewels</a></p>
<h2 id="core-modules--plugins"><a href="#core-modules--plugins" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Core modules / plugins</h2>
<p>New modules can either be written as part of MultiQC or in a stand-alone
plugin. If your module is for a publicly available tool, please add it
to the main program and contribute your code back when complete via a
pull request.</p>
<p>If your module is for something <em>very</em> niche, which no-one else can use,
you can write it as part of a custom plugin. The process is almost identical,
though it keeps the code bases separate. For more information about this,
see the docs about <em>MultiQC Plugins</em> below.</p>
<h2 id="initial-setup"><a href="#initial-setup" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Initial setup</h2>
<h3 id="submodule"><a href="#submodule" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Submodule</h3>
<p>MultiQC modules are Python submodules - as such, they need their own
directory in <code>/multiqc/</code> with an <code>__init__.py</code> file. The directory should
share its name with the module. To follow common practice, the module
code usually then goes in a separate python file (also with the same name)
which is then imported by <code>__init__.py</code>:</p>
<pre><code class="language-python">from __future__ import absolute_import
from .modname import MultiqcModule</code></pre>
<h3 id="entry-points"><a href="#entry-points" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Entry points</h3>
<p>Once your submodule files are in place, you need to tell MultiQC that they
are available as an analysis module. This is done within <code>setup.py</code> using
<a href="https://pythonhosted.org/setuptools/setuptools.html#dynamic-discovery-of-services-and-plugins">entry points</a>.
In <code>setup.py</code> you will see some code that looks like this:</p>
<pre><code class="language-python">entry_points = {
    'multiqc.modules.v1': [
        'bismark = multiqc.modules.bismark:MultiqcModule',
        [...]
    ]
}</code></pre>
<p>Copy one of the existing module lines and change it to use your module name.
The order is irrelevant, so stick to alphabetical if in doubt.
Once this is done, you will need to update your installation of MultiQC:</p>
<pre><code>python setup.py develop</code></pre>
<h3 id="multiqc-config"><a href="#multiqc-config" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC config</h3>
<p>So that MultiQC knows what order modules should be run in, you need to add
your module to the core config file.</p>
<p>In <code>multiqc/utils/config_defaults.yaml</code> you should see a list variable called <code>module_order</code>.
This contains the name of modules in order of precedence. Add your module here
in an appropriate position.</p>
<h3 id="documentation"><a href="#documentation" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Documentation</h3>
<p>Next up, you need to create a documentation file for your module. The reason
for this is twofold: firstly, docs are important to help people to use, debug
and extend MultiQC (you're reading this, aren't you?). Secondly,
having the file there with the appropriate YAML front matter will make the
module show up on the <a href="http://multiqc.info">MultiQC homepage</a> so that everyone
knows it exists. This process is automated once the file is added to the core
repository.</p>
<p>This docs file should be placed in <code>docs/modules/&lt;your_module_name&gt;.md</code> and
should have the following structure:</p>
<pre><code>---
Name: Tool Name
URL: http://www.amazing-bfx-tool.com
Description: &gt;
    This amazing tool does some really cool stuff. You can describe it
    here and split onto multiple lines if you want. Not too long though!
---

Your documentation goes here. Feel free to use markdown and write whatever
you think would be helpful. Please avoid using heading levels 1 to 3.</code></pre>
<h3 id="readme-and-changelog"><a href="#readme-and-changelog" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Readme and Changelog</h3>
<p>Last but not least, remember to add your new module to the main <code>README.md</code>
file and <code>CHANGELOG.md</code>, so that people know that it's there. Feel free to
add your name to the list of credits at the bottom of the readme.</p>
<h3 id="multiqcmodule-class"><a href="#multiqcmodule-class" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiqcModule Class</h3>
<p>If you've copied one of the other entry point statements, it will have
ended in <code>:MultiqcModule</code> - this tells MultiQC to try to execute a class or
function called <code>MultiqcModule</code>.</p>
<p>To use the helper functions bundled with MultiQC, you should extend this
class from <code>multiqc.BaseMultiqcModule</code>. This will give you access to a
number of functions on the <code>self</code> namespace. For example:</p>
<pre><code class="language-python">class MultiqcModule(multiqc.BaseMultiqcModule):
    def __init__(self):
        # Initialise the parent object
        super(MultiqcModule, self).__init__(name='My Module', anchor='mymod',
        href="http://www.awesome_bioinfo.com/my_module",
        info="is an example analysis module used for writing documentation.")</code></pre>
<p>Ok, that should be it! The <code>__init__()</code> function will now be executed every
time MultiQC runs. Try adding a <code>print("Hello World!")</code> statement and see
if it appears in the MultiQC logs at the appropriate time...</p>
<p>Note that the <code>__init__</code> variables are used to create the header, URL link,
analysis module credits and description in the report.</p>
<h3 id="logging"><a href="#logging" class="header-link"><span class="glyphicon glyphicon-link"></span></a> Logging</h3>
<p>Last thing - MultiQC modules have a standardised way of producing output,
so you shouldn't really use <code>print()</code> statements for your <code>Hello World</code> ;)</p>
<p>Instead, use the <code>logger</code> module as follows:</p>
<pre><code class="language-python">import logging
log = logging.getLogger(__name__)
# Initialise your class and so on
log.info('Hello World!')</code></pre>
<p>Log messages can come in a range of formats:</p>
<ul>
<li><code>log.debug</code>
<ul>
<li>Thes only show if MultiQC is run in <code>-v</code>/<code>--verbose</code> mode</li>
</ul></li>
<li><code>log.info</code>
<ul>
<li>For more important status updates</li>
</ul></li>
<li><code>log.warning</code>
<ul>
<li>Alert user about problems that don't halt execution</li>
</ul></li>
<li><code>log.error</code> and <code>log.critical</code>
<ul>
<li>Not often used, these are for show-stopping problems</li>
</ul></li>
</ul>
<h2 id="step-1-find-log-files"><a href="#step-1-find-log-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 1 - Find log files</h2>
<p>The first thing that your module will need to do is to find analysis log
files. You can do this by searching for a filename fragment, or a string
within the file. It's possible to search for both (a match on either
will return the file) and also to have multiple strings possible.</p>
<p>First, add your default patterns to:</p>
<pre><code>MULTIQC_ROOT/multiqc/utils/search_patterns.yaml</code></pre>
<p>You should see the patterns for all other modules to give you an idea,
but you want a yaml key with the name of your module, then either <code>fn</code>
or <code>contents</code> for strings to match against filenames or file contents:</p>
<pre><code class="language-yaml">mymod:
    fn: _myprogram.txt
myothermod:
    contents: This is myprogram v1.3</code></pre>
<p>Note that if you want to find multiple log files, you can nest these
dictionaries (though they must end with either <code>fn</code> or <code>contents</code>).
For example, see the <code>FastQC</code> module:</p>
<pre><code class="language-yaml">fastqc:
    data:
        fn: fastqc_data.txt
    zip:
        fn: _fastqc.zip</code></pre>
<p>You can supply a list of strings if needed, eg. the <code>bismark</code> module:</p>
<pre><code class="language-yaml">bismark:
    align:
        fn:
            - _PE_report.txt
            - _SE_report.txt</code></pre>
<p>The value of adding these strings here is that they can be overwritten
by users in their own config files. This is helpful as people have weird
and wonderful processing pipelines with their own file naming conventions.</p>
<p>Once your strings are added, you can call them in your module with the
<code>config.sp['mymod']</code>. Next, use the base function <code>self.find_log_files()</code>
to look for your files like this:</p>
<pre><code class="language-python">self.find_log_files(config.sp['mymod'], filehandles=False)</code></pre>
<p>This will recursively search the analysis directories looking for a matching
file name (if the <code>fn</code> key is there) or a text string held within a file
(if the <code>contents</code> key is there). Contents matching is only done on files
smaller than <code>config.log_filesize_limit</code> (default 1MB).
Note that both <code>fn</code> and <code>contents</code> can be used in combination
if required - files will be returned if anything matches (<code>OR</code> not <code>AND</code>).</p>
<p>This function yields a dictionary with various information about matching
files. The <code>f</code> key contains the contents of matching files by default.</p>
<pre><code class="language-python"># Find all files for mymod
for f in self.find_log_files(config.sp['mymod']):
    print f['f']        # File contents
    print f['s_name']   # Sample name (from cleaned fn)
    print f['root']     # Directory file was in
    print f['fn']       # Filename</code></pre>
<p>If <code>filehandles=True</code> is specified, the <code>f</code> key contains a file handle
instead:</p>
<pre><code class="language-python"># Find all files which contain the string 'My Statistic:'
# Return a filehandle instead of the file contents
for f in self.find_log_files(config.sp['mymod'], filehandles=True):
    line = f['f'].readline()  # f['f'] is now a filehandle instead of contents</code></pre>
<h2 id="step-2-parse-data-from-the-input-files"><a href="#step-2-parse-data-from-the-input-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 2 - Parse data from the input files</h2>
<p>What most MultiQC modules do once they have found matching analysis files
is to pass the matched file contents to another function, responsible
for parsing the data from the file. How this parsing is done will depend
on the format of the log file and the type of data being read. See below
for a basic example, based loosly on the preseq module:</p>
<pre><code class="language-python">class MultiqcModule(BaseMultiqcModule):
    def __init__(self):
        # [...]
        self.mod_data = dict()
        for f in self.find_log_files(config.sp['mymod']):
            self.mod_data[f['s_name']] = self.parse_logs(f['f'])

    def parse_logs(self, f):
        data = {}
        for l in f.splitlines():
            s = l.split()
            data[s[0]] = s[1]
        return data</code></pre>
<h3 id="no-files-found"><a href="#no-files-found" class="header-link"><span class="glyphicon glyphicon-link"></span></a>No files found</h3>
<p>If your module cannot find any matching files, it needs to raise an
exception of type <code>UserWarning</code>. This tells the core MultiQC program
that no modules were found. For example:</p>
<pre><code class="language-python">if len(self.mod_data) == 0:
    log.debug("Could not find any data in {}".format(config.analysis_dir))
    raise UserWarning</code></pre>
<p>Note that this has to be raised as early as possible, so that it halts
the module progress. For example, if no logs are found then the module
should not create any files or try to do any computation.</p>
<h3 id="custom-sample-names"><a href="#custom-sample-names" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom sample names</h3>
<p>Typically, sample names are taken from cleaned log filenames (the default
<code>f['s_name']</code> value returned). However, if possible, it's better to use
the name of the input file (allowing for concatenated log files).
To do this, you should use the <code>self.clean_s_name()</code> function, as
this will prepend the directory name if requested on the command line:</p>
<pre><code class="language-python">input_fname = s[3] # Or parsed however
s_name = self.clean_s_name(input_fname, f['root'])</code></pre>
<p>This function has already been applied to the contents of <code>f['s_name']</code>.</p>
<blockquote>
<p><code>self.clean_s_name()</code> <strong>must</strong> be used on sample names parsed from the file
contents. Without it, features such as prepending directories (<code>--dirs</code>)
will not work.</p>
</blockquote>
<h3 id="identical-sample-names"><a href="#identical-sample-names" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Identical sample names</h3>
<p>If modules find samples with identical names, then the previous sample
is overwritten. It's good to print a log statement when this happens,
for debugging. However, most of the time it makes sense - programs often
create log files <em>and</em> print to <code>stdout</code> for example.</p>
<pre><code class="language-python">if f['s_name'] in self.bowtie_data:
    log.debug("Duplicate sample name found! Overwriting: {}".format(f['s_name']))</code></pre>
<h3 id="printing-to-the-sources-file"><a href="#printing-to-the-sources-file" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Printing to the sources file</h3>
<p>Finally, once you've found your file we want to add this information to the
<code>multiqc_sources.txt</code> file in the MultiQC report data directory. This lists
every sample name and the file from which this data came from. This is especially
useful if sample names are being overwritten as it lists the source used. This code
is typically written immediately after the above warning.</p>
<p>If you've used the <code>self.find_log_files</code> function, writing to the sources file
is as simple as passing the log file variable to the <code>self.add_data_source</code> function:</p>
<pre><code class="language-python">for f in self.find_log_files(config.sp['mymod']):
    self.add_data_source(f)</code></pre>
<p>If you have different files for different sections of the module, or are
customising the sample name, you can tweak the fields. The default arguments
are as shown:</p>
<pre><code class="language-python">self.add_data_source(f=None, s_name=None, source=None, module=None, section=None)</code></pre>
<h2 id="step-4-adding-to-the-general-statistics-table"><a href="#step-4-adding-to-the-general-statistics-table" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 4 - Adding to the general statistics table</h2>
<p>Now that you have your parsed data, you can start inserting it into the
MultiQC report. At the top of ever report is the 'General Statistics'
table. This contains metrics from all modules, allowing cross-module
comparison.</p>
<p>There is a helper function to add your data to this table. It can take
a lot of configuration options, but most have sensible defaults. At
it's simplest, it works as follows:</p>
<pre><code class="language-python">data = {
    'sample_1': {
        'first_col': 91.4,
        'second_col': '78.2%'
    },
    'sample_2': {
        'first_col': 138.3,
        'second_col': '66.3%'
    }
}
self.general_stats_addcols(data, headers)</code></pre>
<p>To give more informative table headers and configure things like
data scales and colour schemes, you can supply an extra dict:</p>
<pre><code class="language-python">headers = OrderedDict()
headers['first_col'] = {
    'title': 'First',
    'description': 'My First Column',
    'scale': 'RdYlGn-rev'
}
headers['second_col'] = {
    'title': 'Second',
    'description': 'My Second Column',
    'max': 100,
    'min': 0,
    'scale': 'Blues',
    'format': '{:.1f}%'
}
self.general_stats_addcols(data, headers)</code></pre>
<p>Here are all options for headers, with defaults:</p>
<pre><code class="language-python">headers['name'] = {
    'namespace': '',                # Module name. Auto-generated for General Statistics.
    'title': '[ dict key ]',        # Short title, table column title
    'description': '[ dict key ]',  # Longer description, goes in mouse hover text
    'max': None,                    # Minimum value in range, for bar / colour coding
    'min': None,                    # Maximum value in range, for bar / colour coding
    'scale': 'GnBu',                # Colour scale for colour coding
    'format': '{:.1f}',             # Output format() string
    'shared_key': None              # See below for description
    'modify': None,                 # Lambda function to modify values
    'hidden': False                 # Set to True to hide the column on page load
}</code></pre>
<ul>
<li><code>namespace</code>
<ul>
<li>This prepends the column title in the mouse hover: <em>Namespace: Title</em>.
It's automatically generated for the General Statistics table.</li>
</ul></li>
<li><code>scale</code>
<ul>
<li>Colour scales are the names of ColorBrewer palettes. See the
<a href="https://github.com/gka/chroma.js/wiki/Predefined-Colors">chroma.js documentation</a>
for a list of available colour scales</li>
<li>Add <code>-rev</code> to the name of a colour scale to reverse it</li>
</ul></li>
<li><code>shared_key</code>
<ul>
<li>Any string can be specified here, if other columns are found that share
the same key, a consistent colour scheme and data scale will be used in
the table. Typically this is set to things like <code>read_count</code>, so that
the read count in a sample can be seen varying across analysis modules.</li>
</ul></li>
<li><code>modify</code>
<ul>
<li>A python <code>lambda</code> function to change the data in some way when it is
inserted into the table. Typically, this is used to divide numbers to
show millions: <code>'modify': lambda x: x / 1000000</code></li>
</ul></li>
<li><code>hidden</code>
<ul>
<li>Setting this to <code>True</code> will hide the column when the report loads. It can
then be shown through the <em>Configure Columns</em> modal in the report. This can
be useful when data could be sometimes useful. For example, some modules
show &quot;percentage aligned&quot; on page load but hide &quot;number of reads aligned&quot;.</li>
</ul></li>
</ul>
<p>A third parameter can be passed to this function, <code>namespace</code>. This is usually
not needed - MultiQC automatically takes the name of the module that is calling
the function and uses this. However, sometimes it can be useful to overwrite this.</p>
<h2 id="step-5-writing-data-to-a-file"><a href="#step-5-writing-data-to-a-file" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 5 - Writing data to a file</h2>
<p>In addition to printing data to the General Stats, MultiQC modules typically
also write to text-files to allow people to easily use the data in downstream
applications. This also gives the opportunity to output additional data that
may not be appropriate for the General Statistics table.</p>
<p>Again, there is a base class function to help you with this - just supply it
with a dictionary and a filename:</p>
<pre><code class="language-python">data = {
    'sample_1': {
        'first_col': 91.4,
        'second_col': '78.2%'
    },
    'sample_2': {
        'first_col': 138.3,
        'second_col': '66.3%'
    }
}
self.write_data_file(data, 'multiqc_mymod')</code></pre>
<p>If your output has a lot of columns, you can supply the additional
argument <code>sort_cols = True</code> to have the columns alphabetically sorted.</p>
<p>This function will also pay attention to the default / command line
supplied data format and behave accordingly. So the written file could
be a tab-separated file (default), <code>JSON</code> or <code>YAML</code>.</p>
<p>Note that any keys with more than 2 levels of nesting will be ignored
when being written to tab-separated files.</p>
<h2 id="step-6-create-report-sections"><a href="#step-6-create-report-sections" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 6 - Create report sections</h2>
<p>Great! It's time to start creating sections of the report with more information.
If you only have one plot / section to create, just add it to the introduction.
For example (content to be replaced with a brilliant plot in the bit of the docs):</p>
<pre><code class="language-python">self.intro += 'My amazing module output'</code></pre>
<p>If you have multiple plots to show (<em>eg.</em> the Qualimap and FastQC modules),
you can create a list to hold the sections:</p>
<pre><code class="language-python">self.sections = list()
self.sections.append({
    'name': 'First Module Section',
    'anchor': 'mymod_first',
    'content': 'My amazing module output, from the first section'
})
self.sections.append({
    'name': 'Second Module Section',
    'anchor': 'mymod_second',
    'content': 'My amazing module output, from the second section'
})</code></pre>
<p>The will automatically be labelled and linked in the navigation.</p>
<h2 id="step-7-plot-some-data"><a href="#step-7-plot-some-data" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 7 - Plot some data</h2>
<p>Ok, you have some data, now the fun bit - visualising it! Each of the plot
types is described in the <em>Plotting Functions</em> section of the docs.</p>
<h2 id="appendices-1"><a href="#appendices-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Appendices</h2>
<h3 id="adding-custom-css--javascript"><a href="#adding-custom-css--javascript" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Adding Custom CSS / Javascript</h3>
<p>If you would like module-specific CSS and / or JavaScript added to the template,
just add to the <code>self.css</code> and <code>self.js</code> dictionaries that come with the
<code>BaseMultiqcModule</code> class. The key should be the filename that you want your file to
have in the generated report folder <em>(this is ignored in the default template, which
includes the content file directly in the HTML)</em>. The dictionary value should be
the path to the desired file. For example, see how it's done in the FastQC module:</p>
<pre><code class="language-python">self.css = {
    'assets/css/multiqc_fastqc.css' :
        os.path.join(os.path.dirname(__file__), 'assets', 'css', 'multiqc_fastqc.css')
}
self.js = {
    'assets/js/multiqc_fastqc.js' :
        os.path.join(os.path.dirname(__file__), 'assets', 'js', 'multiqc_fastqc.js')
}</code></pre></div><div class="docs_block" id="plots.md"><h1 id="plotting-functions"><a href="#plotting-functions" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Plotting Functions</h1>
<p>MultiQC plotting functions are held within <code>multiqc.plots</code> submodules.
To use them, simply import the modules you want, eg.:</p>
<pre><code class="language-python">from multiqc.plots import bargraph, linegraph</code></pre>
<p>Once you've done that, you will have access to the corresponding plotting
functions:</p>
<pre><code class="language-python">bargraph.plot()
linegraph.plot()
scatter.plot()
table.plot()
beeswarm.plot()
heatmap.plot()</code></pre>
<p>These have been designed to work in a similar manner to each other - you
pass a data structure to them, along with optional extras such as categories
and configuration options, and they return a string of HTML to add to the
report. You can add this to the module introduction or sections as described
above. For example:</p>
<pre><code class="language-python">self.sections.append({
    'name': 'Module Section',
    'anchor': 'mymod_section',
    'content': bargraph.plot(self.parsed_data, categories, pconfig)
})</code></pre>
<h2 id="bar-graphs"><a href="#bar-graphs" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bar graphs</h2>
<p>Simple data can be plotted in bar graphs. Many MultiQC modules make use
of stacked bar graphs. Here, the <code>bargraph.plot()</code> function comes to
the rescue. A basic example is as follows:</p>
<pre><code class="language-python">from multiqc import plots
data = {
    'sample 1': {
        'aligned': 23542,
        'not_aligned': 343,
    },
    'sample 2': {
        'not_aligned': 7328,
        'aligned': 1275,
    }
}
html_content = bargraph.plot(data)</code></pre>
<p>To specify the order of categories in the plot, you can supply a list of
dictionary keys. This can also be used to exclude a key from the plot.</p>
<pre><code class="language-python">cats = ['aligned', 'not_aligned']
html_content = bargraph.plot(data, cats)</code></pre>
<p>If <code>cats</code> is given as a dict instead of a list, you can specify a nice name
and a colour too. Make it an OrderedDict to specify the order:</p>
<pre><code class="language-python">from collections import OrderedDict
cats = OrderedDict()
cats['aligned'] = {
    'name': 'Aligned Reads',
    'color': '#8bbc21'
}
cats['not_aligned'] = {
    'name': 'Unaligned Reads',
    'color': '#f7a35c'
}</code></pre>
<p>Finally, a third variable can be supplied with configuration variables for
the plot. The defaults are as follows:</p>
<pre><code class="language-python">config = {
    # Building the plot
    'id': '&lt;random string&gt;',                # HTML ID used for plot
    'cpswitch': True,                       # Show the 'Counts / Percentages' switch?
    'cpswitch_c_active': True,              # Initial display with 'Counts' specified? False for percentages.
    'cpswitch_counts_label': 'Counts',      # Label for 'Counts' button
    'cpswitch_percent_label': 'Percentages' # Label for 'Percentages' button
    'logswitch': False,                     # Show the 'Log10' switch?
    'logswitch_active': False,              # Initial display with 'Log10' active?
    'logswitch_label': 'Log10',             # Label for 'Log10' button
    'hide_zero_cats': True,                 # Hide categories where data for all samples is 0
    # Customising the plot
    'title': None,                          # Plot title
    'xlab': None,                           # X axis label
    'ylab': None,                           # Y axis label
    'ymax': None,                           # Max y limit
    'ymin': None,                           # Min y limit
    'yCeiling': None,                       # Maximum value for automatic axis limit (good for percentages)
    'yFloor': None,                         # Minimum value for automatic axis limit
    'yMinRange': None,                      # Minimum range for axis
    'yDecimals': True,                      # Set to false to only show integer labels
    'ylab_format': None,                    # Format string for x axis labels. Defaults to {value}
    'stacking': 'normal',                   # Set to None to have category bars side by side
    'use_legend': True,                     # Show / hide the legend
    'click_func': None,                     # Javascript function to be called when a point is clicked
    'cursor': None,                         # CSS mouse cursor type.
    'tt_decimals': 0,                       # Number of decimal places to use in the tooltip number
    'tt_suffix': '',                        # Suffix to add after tooltip number
    'tt_percentages': True,                 # Show the percentages of each count in the tooltip
}</code></pre>
<h3 id="switching-datasets"><a href="#switching-datasets" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Switching datasets</h3>
<p>It's possible to have single plot with buttons to switch between different
datasets. To do this, give a list of data objects (same formats as described
above). Also add the following config options to supply names to the buttons:</p>
<pre><code class="language-python">config = {
    'data_labels': ['Reads', 'Bases']
}</code></pre>
<p>You can also customise the y-axis label and min/max values for each dataset:</p>
<pre><code class="language-python">config = {
    'data_labels': [
        {'name': 'Reads', 'ylab': 'Number of Reads'},
        {'name': 'Bases', 'ylab': 'Number of Base Pairs', 'ymax':100}
    ]
}</code></pre>
<p>If supplying multiple datasets, you can also supply a list of category
objects. Make sure that they are in the same order as the data.</p>
<p>Categories should contain data keys, so if you're supplying a list of two datasets,
you should supply a list of two sets of keys for the categories. MultiQC will try to
guess categories from the data keys if categories are missing.</p>
<p>For example, with two datasets supplied as above:</p>
<pre><code class="language-python">cats = [
    ['aligned_reads','unaligned_reads'],
    ['aligned_base_pairs','unaligned_base_pairs'],
]</code></pre>
<p>Or with additional customisation such as name and colour:</p>
<pre><code class="language-python">from collections import OrderedDict
cats = [OrderedDict(), OrderedDict()]
cats[0]['aligned_reads'] =        {'name': 'Aligned Reads',        'color': '#8bbc21'}
cats[0]['unaligned_reads'] =      {'name': 'Unaligned Reads',      'color': '#f7a35c'}
cats[1]['aligned_base_pairs'] =   {'name': 'Aligned Base Pairs',   'color': '#8bbc21'}
cats[1]['unaligned_base_pairs'] = {'name': 'Unaligned Base Pairs', 'color': '#f7a35c'}</code></pre>
<h3 id="interactive--flat-image-plots"><a href="#interactive--flat-image-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Interactive / Flat image plots</h3>
<p>Note that the <code>bargraph.plot()</code> function can generate both interactive
JavaScript (HighCharts) powered report plots <em>and</em> flat image plots made using
MatPlotLib. This choice is made within the function based on config variables
such as number of dataseries and command line flags.</p>
<p>Note that both plot types should come out looking pretty much identical. If
you spot something that's missing in the flat image plots, let me know.</p>
<h2 id="line-graphs"><a href="#line-graphs" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Line graphs</h2>
<p>This base function works much like the above, but for two-dimensional
data, to produce line graphs. It expects a dictionary in the following format:</p>
<pre><code class="language-python">from multiqc import plots
data = {
    'sample 1': {
        '&lt;x val 1&gt;': '&lt;y val 1&gt;',
        '&lt;x val 2&gt;': '&lt;y val 2&gt;',
    },
    'sample 2': {
        '&lt;x val 1&gt;': '&lt;y val 1&gt;',
        '&lt;x val 2&gt;': '&lt;y val 2&gt;',
    }
}
html_content = linegraph.plot(data)</code></pre>
<p>Additionally, a config dict can be supplied. The defaults are as follows:</p>
<pre><code class="language-python">from multiqc import plots
config = {
    # Building the plot
    'smooth_points': None,       # Supply a number to limit number of points / smooth data
    'smooth_points_sumcounts': True, # Sum counts in bins, or average? Can supply list for multiple datasets
    'id': '&lt;random string&gt;',     # HTML ID used for plot
    'categories': False,         # Set to True to use x values as categories instead of numbers.
    'colors': dict()             # Provide dict with keys = sample names and values colours
    'extra_series': None,        # See section below
    # Plot configuration
    'title': None,               # Plot title
    'xlab': None,                # X axis label
    'ylab': None,                # Y axis label
    'xCeiling': None,            # Maximum value for automatic axis limit (good for percentages)
    'xFloor': None,              # Minimum value for automatic axis limit
    'xMinRange': None,           # Minimum range for axis
    'xmax': None,                # Max x limit
    'xmin': None,                # Min x limit
    'xLog': False,               # Use log10 x axis?
    'xDecimals': True,           # Set to false to only show integer labels
    'yCeiling': None,            # Maximum value for automatic axis limit (good for percentages)
    'yFloor': None,              # Minimum value for automatic axis limit
    'yMinRange': None,           # Minimum range for axis
    'ymax': None,                # Max y limit
    'ymin': None,                # Min y limit
    'yLog': False,               # Use log10 y axis?
    'yDecimals': True,           # Set to false to only show integer labels
    'yPlotBands': None,          # Highlighted background bands. See http://api.highcharts.com/highcharts#yAxis.plotBands
    'xPlotBands': None,          # Highlighted background bands. See http://api.highcharts.com/highcharts#xAxis.plotBands
    'yPlotLines': None,          # Highlighted background lines. See http://api.highcharts.com/highcharts#yAxis.plotLines
    'xPlotLines': None,          # Highlighted background lines. See http://api.highcharts.com/highcharts#xAxis.plotLines
    'tt_label': '{point.x}: {point.y:.2f}', # Use to customise tooltip label, eg. '{point.x} base pairs'
    'pointFormat': None,         # Replace the default HTML for the entire tooltip label
    'click_func': function(){},  # Javascript function to be called when a point is clicked
    'cursor': None               # CSS mouse cursor type. Defaults to pointer when 'click_func' specified
    'reversedStacks': False      # Reverse the order of the category stacks. Defaults True for plots with Log10 option
}
html_content = linegraph.plot(data, config)</code></pre>
<h3 id="switching-datasets-1"><a href="#switching-datasets-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Switching datasets</h3>
<p>You can also have a single plot with buttons to switch between different
datasets. To do this, just supply a list of data dicts instead (same
formats as described above). Also add the following config options to
supply names to the buttons and graph labels:</p>
<pre><code class="language-python">config = {
    'data_labels': [
        {'name': 'DS 1', 'ylab': 'Dataset 1'},
        {'name': 'DS 2', 'ylab': 'Dataset 2'}
    ]
}</code></pre>
<p>All of these config values are optional, the function will default
to sensible values if things are missing. See the cutadapt module
plots for an example of this in action.</p>
<h3 id="additional-data-series"><a href="#additional-data-series" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Additional data series</h3>
<p>Sometimes, it's good to be able to specify specific data series manually.
To do this, use <code>config['extra_series']</code>. For a single extra line this can
be a dict (as below). For multiple lines, use a list of dicts. For multiple
dataset plots, use a list of list of dicts.</p>
<p>For example, to add a dotted <code>x = y</code> reference line:</p>
<pre><code class="language-python">from multiqc import plots
config = {
    'extra_series': {
        'name': 'x = y',
        'data': [[0, 0], [max_x_val, max_y_val]],
        'dashStyle': 'Dash',
        'lineWidth': 1,
        'color': '#000000',
        'marker': { 'enabled': False },
        'enableMouseTracking': False,
        'showInLegend': False,
    }
}
html_content = linegraph.plot(data, config)</code></pre>
<h2 id="scatter-plots"><a href="#scatter-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Scatter Plots</h2>
<p>Scatter plots work in almost exactly the same way as line plots. Most (if not all)
config options are shared between the two. The data structure is similar but
not identical:</p>
<pre><code class="language-python">from multiqc import plots
data = {
    'sample 1': {
        x: '&lt;x val&gt;',
        y: '&lt;y val&gt;'
    },
    'sample 2': {
        x: '&lt;x val&gt;',
        y: '&lt;y val&gt;'
    }
}
html_content = scatter.plot(data)</code></pre>
<p>If you want more than one data point per sample, you can supply a list of
dictionaries instead. You can also optionally specify point colours and
sample name suffixes (these are appended to the sample name):</p>
<pre><code class="language-python">data = {
    'sample 1': [
        { x: '&lt;x val&gt;', y: '&lt;y val&gt;', color: '#a6cee3', name: 'Type 1' },
        { x: '&lt;x val&gt;', y: '&lt;y val&gt;', color: '#1f78b4', name: 'Type 2' }
    ],
    'sample 2': [
        { x: '&lt;x val&gt;', y: '&lt;y val&gt;', color: '#b2df8a', name: 'Type 1' },
        { x: '&lt;x val&gt;', y: '&lt;y val&gt;', color: '#33a02c', name: 'Type 2' }
    ]
}</code></pre>
<p>Remember that MultiQC reports can contain large numbers of samples, so this plot type
is <strong>not</strong> suitable for large quantities of data - 20,000 genes might look good
for one sample, but when someone runs MultiQC with 500 samples, it will crash
the browser and be impossible to interpret.</p>
<p>See the above docs about line plots for most config options. The scatter plot
has a handful of unique ones in addition:</p>
<pre><code class="language-python">pconfig = {
    'marker_colour': 'rgba(124, 181, 236, .5)', # string, base colour of points (recommend rgba / semi-transparent)
    'marker_size': 5,               # int, size of points
    'marker_line_colour': '#999',   # string, colour of point border
    'marker_line_width': 1,         # int, width of point border
    'square': False                 # Force the plot to stay square? (Maintain aspect ratio)
}</code></pre>
<h2 id="creating-a-table"><a href="#creating-a-table" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Creating a table</h2>
<p>Tables should work just like the functions above (most like the bar
graph function). As a minimum, the function takes a dictionary containing
data - the first keys will be sample names (row headers) and each key
contained within will be a table column header.</p>
<p>You can also supply a list of key names to restrict the data in the table
to certain keys / columns. This also specifies the order that columns
should be displayed in.</p>
<p>For more customisation, the headers can be supplied as a dictionary. Each
key should match the keys used in the data dictionary, but values can
customise the output. If you want to specify the order of the columns, you
must use an <code>OrderedDict</code>.</p>
<p>Finally, a the function accepts a third parameter, a config dictionary.
This can set global options for the table (eg. a title) and can also hold
default values to customise the output of all table columns.</p>
<p>The default header keys and table config options are:</p>
<pre><code class="language-python">single_header = {
    'namespace': '',                # Name for grouping in table
    'title': '[ dict key ]',        # Short title, table column title
    'description': '[ dict key ]',  # Longer description, goes in mouse hover text
    'max': None,                    # Minimum value in range, for bar / colour coding
    'min': None,                    # Maximum value in range, for bar / colour coding
    'scale': 'GnBu',                # Colour scale for colour coding
    'colour': '&lt;auto from palette&gt;',# Colour for column grouping
    'format': '{:.1f}',             # Output format() string
    'shared_key': None              # See below for description
    'modify': None,                 # Lambda function to modify values
    'hidden': False                 # Set to True to hide the column on page load
}
table_config = {
    'id': '&lt;random string&gt;',                 # ID used for the table
    'table_title': '&lt;table id&gt;',             # Title of the table. Used in the column config modal
    'save_file': False,                      # Whether to save the table data to a file
    'raw_data_fn':'multiqc_&lt;table_id&gt;_table' # File basename to use for raw data file
    'no_beeswarm': False    # Force a table to always be plotted (beeswarm by default if many rows)
}</code></pre>
<p>Finally, a third parameter can be specified with table settings.</p>
<pre><code class="language-python">tconfig = {
    'id': None,             # ID for the table in the HTML
    'table_title': None,    # Title printed above the table
    'save_file': False,     # Save the data in the table to a file in `multiqc_data`
    'raw_data_fn': None,    # Filename to use if saving data file
    'no_beeswarm': False    # Force a table to always be plotted (beeswarm by default if many rows)
}</code></pre>
<p>A very basic example is shown below:</p>
<pre><code class="language-python">data = {
    'sample 1': {
        'aligned': 23542,
        'not_aligned': 343,
    },
    'sample 2': {
        'aligned': 1275,
        'not_aligned': 7328,
    }
}
table_html = table.plot(data)</code></pre>
<p>A more complicated version with ordered columns, defaults and column-specific
settings:</p>
<pre><code class="language-python">data = {
    'sample 1': {
        'aligned': 23542,
        'not_aligned': 343,
        'aligned_percent': 98.563952271
    },
    'sample 2': {
        'aligned': 1275,
        'not_aligned': 7328,
        'aligned_percent': 14.820411484
    }
}
headers = OrderedDict()
headers['aligned_percent'] = {
    'title': '% Aligned',
    'description': 'Percentage of reads that aligned',
    'format': '{:.1f}%',
    'suffix': '%',
    'max': 100,
}
headers['aligned'] = {
    'title': 'M Aligned',
    'description': 'Aligned Reads (millions)',
    'format': '{:.2f}',
    'shared_key': 'read_count',
    'modify': lambda x: x / 1000000
}
config = {
    'namespace': 'My Module',
    'min': 0,
    'scale': 'GnBu'
}
table_html = table.plot(data, headers, config)</code></pre>
<h2 id="beeswarm-plots-dot-plots"><a href="#beeswarm-plots-dot-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Beeswarm plots (dot plots)</h2>
<p>Beeswarm plots work from the exact same data structure as tables, so the
usage is just the same. Except instead of calling <code>table</code>, call <code>beeswarm</code>:</p>
<pre><code class="language-python">data = {
    'sample 1': {
        'aligned': 23542,
        'not_aligned': 343,
    },
    'sample 2': {
        'not_aligned': 7328,
        'aligned': 1275,
    }
}
beeswarm_html = beeswarm.plot(data)</code></pre>
<p>The function also accepts the same headers and config parameters.</p>
<h2 id="heatmaps"><a href="#heatmaps" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Heatmaps</h2>
<p>Heatmaps expect data in the structure of a list of lists. Then, a list
of sample names for the x-axis, and optionally for the y-axis (defaults
to the same as the x-axis).</p>
<pre><code class="language-python">heatmap.plot(data, xcats, ycats, pconfig)</code></pre>
<p>A simple example:</p>
<pre><code class="language-python">hmdata = [
    [0.9, 0.87, 0.73, 0.6, 0.2, 0.3],
    [0.87, 1, 0.7, 0.6, 0.9, 0.3],
    [0.73, 0.8, 1, 0.6, 0.9, 0.3],
    [0.6, 0.8, 0.7, 1, 0.9, 0.3],
    [0.2, 0.8, 0.7, 0.6, 1, 0.3],
    [0.3, 0.8, 0.7, 0.6, 0.9, 1],
]
names = [ 'one', 'two', 'three', 'four', 'five', 'six' ]
hm_html = heatmap.plot(hmdata, names)</code></pre>
<p>Much like the other plots, you can change the way that the heatmap looks
using a config dictionary:</p>
<pre><code class="language-python">pconfig = {
    'title': None,                 # Plot title
    'xTitle': None,                # X-axis title
    'yTitle': None,                # Y-axis title
    'min': None,                   # Minimum value (default: auto)
    'max': None,                   # Maximum value (default: auto)
    'square': True,                # Force the plot to stay square? (Maintain aspect ratio)
    'colstops': []                 # Scale colour stops. See below.
    'reverseColors': False,        # Reverse the order of the colour axis
    'decimalPlaces': 2,            # Number of decimal places for tooltip
    'legend': True,                # Colour axis key enabled or not
    'borderWidth': 0,              # Border width between cells
    'datalabels': True,            # Show values in each cell. Defaults True when less than 20 samples.
    'datalabel_colour': '&lt;auto&gt;',  # Colour of text for values. Defaults to auto contrast.
}</code></pre>
<p>The colour stops are a bit special and can be used to define a custom colour
scheme. These should be defined as a list of lists, with a number between 0 and 1
and a HTML colour. The default is <code>RdYlBu</code> from <a href="http://colorbrewer2.org/">ColorBrewer</a>:</p>
<pre><code class="language-python">pconfig = {
    'colstops' = [
        [0, '#313695'],
        [0.1, '#4575b4'],
        [0.2, '#74add1'],
        [0.3, '#abd9e9'],
        [0.4, '#e0f3f8'],
        [0.5, '#ffffbf'],
        [0.6, '#fee090'],
        [0.7, '#fdae61'],
        [0.8, '#f46d43'],
        [0.9, '#d73027'],
        [1, '#a50026'],
    ]
}</code></pre>
<h2 id="javascript-functions"><a href="#javascript-functions" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Javascript Functions</h2>
<p>The javascript bundled in the default MultiQC template has a number of
helper functions to make your life easier.</p>
<blockquote>
<p>NB: The MultiQC Python functions make use of these, so it's very unlikely
that you'll need to use any of this. But it's here for reference.</p>
</blockquote>
<h3 id="plotting-line-graphs"><a href="#plotting-line-graphs" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Plotting line graphs</h3>
<p><code>plot_xy_line_graph (target, ds)</code></p>
<p>Plots a line graph with multiple series of (x,y) data pairs. Used by
the <a href="http://multiqc.info/docs/#line-graphs">linegraph.plot()</a>
python function.</p>
<p>Data and configuration must be added to the document level
<code>mqc_plots</code> variable on page load, using the target as the key.
The variables used are as follows:</p>
<pre><code class="language-javascript">mqc_plots[target]['plot_type'] = 'xy_line';
mqc_plots[target]['config'];
mqc_plots[target]['datasets'];</code></pre>
<p>Multiple datasets can be added in the <code>['datasets']</code> array. The supplied
variable <code>ds</code> specifies which is plotted (defaults to <code>0</code>).</p>
<p>Available config options with default vars:</p>
<pre><code class="language-javascript">config = {
    title: undefined,            // Plot title
    xlab: undefined,             // X axis label
    ylab: undefined,             // Y axis label
    xCeiling: undefined,         // Maximum value for automatic axis limit (good for percentages)
    xFloor: undefined,           // Minimum value for automatic axis limit
    xMinRange: undefined,        // Minimum range for axis
    xmax: undefined,             // Max x limit
    xmin: undefined,             // Min x limit
    xDecimals: true,             // Set to false to only show integer labels
    yCeiling: undefined,         // Maximum value for automatic axis limit (good for percentages)
    yFloor: undefined,           // Minimum value for automatic axis limit
    yMinRange: undefined,        // Minimum range for axis
    ymax: undefined,             // Max y limit
    ymin: undefined,             // Min y limit
    yDecimals: true,             // Set to false to only show integer labels
    yPlotBands: undefined,       // Highlighted background bands. See http://api.highcharts.com/highcharts#yAxis.plotBands
    xPlotBands: undefined,       // Highlighted background bands. See http://api.highcharts.com/highcharts#xAxis.plotBands
    tt_label: '{point.x}: {point.y:.2f}', // Use to customise tooltip label, eg. '{point.x} base pairs'
    pointFormat: undefined,      // Replace the default HTML for the entire tooltip label
    click_func: function(){},    // Javascript function to be called when a point is clicked
    cursor: undefined            // CSS mouse cursor type. Defaults to pointer when 'click_func' specified
}</code></pre>
<p>An example of the markup expected, with the function being called:</p>
<pre><code class="language-html">&lt;div id="my_awesome_line_graph" class="hc-plot"&gt;&lt;/div&gt;
&lt;script type="text/javascript"&gt;
    mqc_plots['#my_awesome_bar_plot']['plot_type'] = 'xy_line';
    mqc_plots['#my_awesome_line_graph']['datasets'] = [
        {
            name: 'Sample 1',
            data: [[1, 1.5], [1.5, 3.1], [2, 6.4]]
        },
        {
            name: 'Sample 2',
            data: [[1, 1.7], [1.5, 4.3], [2, 8.4]]
        },
    ];
    mqc_plots['#my_awesome_line_graph']['config'] = {
        "title": "Best Plot Ever",
        "ylab": "Pings",
        "xlab": "Pongs"
    };
    $(function () {
        plot_xy_line_graph('#my_awesome_line_graph');
    });
&lt;/script&gt;</code></pre>
<h3 id="plotting-bar-graphs"><a href="#plotting-bar-graphs" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Plotting bar graphs</h3>
<p><code>plot_stacked_bar_graph (target, ds)</code></p>
<p>Plots a bar graph with multiple series containing multiple categories.
Used by the <a href="http://multiqc.info/docs/#bar-graphs">bargraph.plot()</a>
python function.</p>
<p>Data and configuration must be added to the document level
<code>mqc_plots</code> variable on page load, using the target as the key.
The variables used are as follows:</p>
<pre><code class="language-javascript">mqc_plots[target]['plot_type'] = 'bar_graph';
mqc_plots[target]['config'];
mqc_plots[target]['datasets'];
mqc_plots[target]['samples'];</code></pre>
<p>All available config options with default vars:</p>
<pre><code class="language-javascript">config = {
    title: undefined,           // Plot title
    xlab: undefined,            // X axis label
    ylab: undefined,            // Y axis label
    ymax: undefined,            // Max y limit
    ymin: undefined,            // Min y limit
    yDecimals: true,            // Set to false to only show integer labels
    ylab_format: undefined,     // Format string for x axis labels. Defaults to {value}
    stacking: 'normal',         // Set to null to have category bars side by side (None in python)
    xtype: 'linear',            // Axis type. 'linear' or 'logarithmic'
    use_legend: true,           // Show / hide the legend
    click_func: undefined,      // Javascript function to be called when a point is clicked
    cursor: undefined,          // CSS mouse cursor type. Defaults to pointer when 'click_func' specified
    tt_percentages: true,       // Show the percentages of each count in the tooltip
    reversedStacks: false,      // Reverse the order of the categories in the stack.
}</code></pre>
<p>An example of the markup expected, with the function being called:</p>
<pre><code class="language-html">&lt;div id="my_awesome_bar_plot" class="hc-plot"&gt;&lt;/div&gt;
&lt;script type="text/javascript"&gt;
    mqc_plots['#my_awesome_bar_plot']['plot_type'] = 'bar_graph';
    mqc_plots['#my_awesome_bar_plot']['samples'] = ['Sample 1', 'Sample 2']
    mqc_plots['#my_awesome_bar_plot']['datasets'] = [{"data": [4, 7], "name": "Passed Test"}, {"data": [2, 3], "name": "Failed Test"}]
    mqc_plots['#my_awesome_bar_plot']['config'] = {
        "title": "My Awesome Plot",
        "ylab": "# Observations",
        "ymin": 0,
        "stacking": "normal"
    };
    $(function () {
        plot_stacked_bar_graph("#my_awesome_bar_plot");
    });
&lt;/script&gt;</code></pre>
<h3 id="switching-counts-and-percentages"><a href="#switching-counts-and-percentages" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Switching counts and percentages</h3>
<p>If you're using the plotting functions above, it's easy to add a button which
switches between percentages and counts. Just add the following HTML above
your plot:</p>
<pre><code class="language-html">&lt;div class="btn-group switch_group"&gt;
    &lt;button class="btn btn-default btn-sm active" data-action="set_numbers" data-target="#my_plot"&gt;Counts&lt;/button&gt;
    &lt;button class="btn btn-default btn-sm" data-action="set_percent" data-target="#my_plot"&gt;Percentages&lt;/button&gt;
&lt;/div&gt;</code></pre>
<p><em>NB:</em> This markup is generated automatically with the Python <code>self.plot_bargraph()</code> function.</p>
<h3 id="switching-plot-datasets"><a href="#switching-plot-datasets" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Switching plot datasets</h3>
<p>Much like the counts / percentages buttons above, you can add a button which
switches the data displayed in a single plot. Make sure that both datasets
are stored in named javascript variables, then add the following markup:</p>
<pre><code class="language-html">&lt;div class="btn-group switch_group"&gt;
    &lt;button class="btn btn-default btn-sm active" data-action="set_data" data-ylab="First Data" data-newdata="data_var_1" data-target="#my_plot"&gt;Data 1&lt;/button&gt;
    &lt;button class="btn btn-default btn-sm" data-action="set_data" data-ylab="Second Data" data-newdata="data_var_2" data-target="#my_plot"&gt;Data 2&lt;/button&gt;
&lt;/div&gt;</code></pre>
<p>Note the CSS class <code>active</code> which specifies which button is 'pressed' on page load.
<code>data-ylab</code> and <code>data-xlab</code> can be used to specify the new axes labels.
<code>data-newdata</code> should be the name of the javascript object with the new data
to be plotted and <code>data-target</code> should be the CSS selector of the plot to change.</p>
<h3 id="custom-event-triggers"><a href="#custom-event-triggers" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom event triggers</h3>
<p>Some of the events that take place in the general javascript
code trigger jQuery events which you can hook into from within your
module's code. This allows you to take advantage of events generated
by the global theme whilst keeping your code modular.</p>
<pre><code class="language-javascript">$(document).on('mqc_highlights', function(e, f_texts, f_cols, regex_mode){
    // This trigger is called when the highlight strings are
    // updated. Three variables are given - an array of search
    // strings (f_texts), an array of colours with corresponding
    // indexes (f_cols) and a boolean var saying whether the
    // search should be treated as a string or a regex (regex_mode)
});

$(document).on('mqc_renamesamples', function(e, f_texts, t_texts, regex_mode){
    // This trigger is called when samples are renamed
    // Three variables are given - an array of search
    // strings (f_texts), an array of replacements with corresponding
    // indexes (t_texts) and a boolean var saying whether the
    // search should be treated as a string or a regex (regex_mode)
});

$(document).on('mqc_hidesamples', function(e, f_texts, regex_mode){
    // This trigger is called when the Hide Samples filters change.
    // Two variables are given - an array of search strings
    // (f_texts) and a boolean saying whether the search should
    // be treated as a string or a regex (regex_mode)
});

$('#YOUR_PLOT_ID').on('mqc_plotresize', function(){
    // This trigger is called when a plot handle is pulled,
    // resizing the height
});

$('#YOUR_PLOT_ID').on('mqc_original_series_click', function(e, name){
    // A plot able to show original images has had a point clicked.
    // 'name' contains the name of the series that was clicked
});

$('#YOUR_PLOT_ID').on('mqc_original_chg_source', function(e, name){
    // A plot with original images has had a request to change the
    // original image source (eg. pressing Prev / Next)
});</code></pre></div><div class="docs_block" id="plugins.md"><h1 id="multiqc-plugins"><a href="#multiqc-plugins" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC Plugins</h1>
<p>MultiQC is written around a system designed for extensibility and plugins.
These features allow custom code to be written without polluting the central
code base.</p>
<p>Please note that we want MultiQC to grow as a community tool! So if you're
writing a module or theme that can be used by others, please keep it within
the main MultiQC framework and submit a pull request.</p>
<h2 id="entry-points-1"><a href="#entry-points-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Entry Points</h2>
<p>The plugin system works using setuptools
<a href="https://pythonhosted.org/setuptools/setuptools.html#dynamic-discovery-of-services-and-plugins">entry points</a>.
In <code>setup.py</code> you will see a section of code that looks like this <em>(truncated)</em>:</p>
<pre><code class="language-python">entry_points = {
    'multiqc.modules.v1': [
        'qualimap = multiqc.modules.qualimap:MultiqcModule',
    ],
    'multiqc.templates.v1': [
        'default = multiqc.templates.default',
    ],
    # 'multiqc.cli_options.v1': [
        # 'my-new-option = myplugin.cli:new_option'
    # ],
    # 'multiqc.hooks.v1': [
        # 'execution_start = myplugin.hooks:execution_start',
        # 'config_loaded = myplugin.hooks:config_loaded',
        # 'before_modules = myplugin.hooks:before_modules',
        # 'after_modules = myplugin.hooks:after_modules',
        # 'execution_finish = myplugin.hooks:execution_finish',
    # ]
},</code></pre>
<p>These sets of entry points can each be extended to add functionality
to MultiQC:</p>
<ul>
<li><code>multiqc.modules.v1</code>
<ul>
<li>Defines the module classes. Used to add new modules.</li>
</ul></li>
<li><code>multiqc.templates.v1</code>
<ul>
<li>Defines the templates. Can be used for new templates.</li>
</ul></li>
<li><code>multiqc.cli_options.v1</code>
<ul>
<li>Allows plugins to add new custom command line options</li>
</ul></li>
<li><code>multiqc.hooks.v1</code>
<ul>
<li>Code hooks for plugins to add new functionality</li>
</ul></li>
</ul>
<p>Any python program can create entry points with the same name, once installed
MultiQC will find these and run them accordingly. For an example of this in
action, see the <a href="https://github.com/ewels/MultiQC_NGI/blob/master/setup.py">MultiQC_NGI</a>
setup file:</p>
<pre><code class="language-python">entry_points = {
        'multiqc.templates.v1': [
            'ngi = multiqc_ngi.templates.ngi',
            'genstat = multiqc_ngi.templates.genstat',
        ],
        'multiqc.cli_options.v1': [
            'project = multiqc_ngi.cli:pid_option'
        ],
        'multiqc.hooks.v1': [
            'after_modules = multiqc_ngi.hooks:ngi_metadata',
        ]
    },</code></pre>
<p>Here, two new templates are added, a new command line option and a new code hook.</p>
<h2 id="modules"><a href="#modules" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Modules</h2>
<p>List items added to <code>multiqc.modules.v1</code> specify new modules. They should
be described as follows:</p>
<pre><code>modname = python_mod.dirname.submodname:classname'</code></pre>
<p>Once this is done, everything else should be the same as described in the
<a href="http://multiqc.info/docs/#writing-new-modules">writing modules</a> documentation.</p>
<h2 id="templates"><a href="#templates" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Templates</h2>
<p>As above, though no need to specify a class name at the end. See the
<a href="http://multiqc.info/docs/#writing-new-templates">writing templates</a> documentation for further instructions.</p>
<h2 id="command-line-options"><a href="#command-line-options" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Command line options</h2>
<p>MultiQC handles command line interaction using the <a href="http://click.pocoo.org/">click</a>
framework. You can use the <code>multiqc.cli_options.v1</code> entry point to add new
click decorators for command line options. For example, the MultiQC_NGI
plugin uses the entry point above with the following code in <code>cli.py</code>:</p>
<pre><code class="language-python">import click
pid_option = click.option('--project', type=str)</code></pre>
<p>The values given from additional command line arguments are parsed by
MultiQC and put into <code>config.kwargs</code>. The above plugin later reads
the value given by the user with the <code>--project</code> flag in a hook:</p>
<pre><code class="language-python">if config.kwargs['project'] is not None:
  # do some stuff</code></pre>
<p>See the <a href="http://click.pocoo.org/">click documentation</a> or the main
MultiQC script for more information and examples of adding command line
options.</p>
<h2 id="hooks"><a href="#hooks" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Hooks</h2>
<p>Hooks are a little more complicated - these define points in the core
MultiQC code where you can run custom functions. This can be useful as
your code is able to access data generated by other parts of the program.
For example, you could tie into the <code>after_modules</code> hook to insert data
processed by MultiQC modules into a database automatically.</p>
<p>Here, the entry point names are the hook titles, described as commented out
lines in the core MultiQC <code>setup.py</code>: <code>execution_start</code>, <code>config_loaded</code>,
<code>before_modules</code>, <code>after_modules</code> and <code>execution_finish</code>.</p>
<p>These should point to a function in your code which will be executed when
that hook fires. Your custom code can import the core MultiQC modules to
access configuration and loggers. For example:</p>
<pre><code class="language-python">#!/usr/bin/env python
""" MultiQC hook functions - we tie into the MultiQC
core here to add in extra functionality. """

import logging
from multiqc.utils import report, config

log = logging.getLogger('multiqc')

def after_modules():
  """ Plugin code to run when MultiQC modules have completed  """
  num_modules = len(report.modules_output)
  status_string = "MultiQC hook - {} modules reported!".format(num_modules)
  log.critical(status_string)</code></pre></div></div>        </div>
        <div class="col-sm-3" id="toc_column">
          <div id="toc" data-spy="affix" data-offset-top="254">
            
<ul class="nav nav-stacked">
<li><a href="#using-multiqc">Using MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#installing-multiqc">Installing MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#installing-python">Installing Python</a></li>
<li><a href="#installing-with-conda">Installing with conda</a></li>
<li><a href="#installation-with-pip">Installation with pip</a></li>
<li><a href="#manual-installation">Manual installation</a></li>
<li><a href="#updating-multiqc">Updating MultiQC</a></li>
<li><a href="#installing-as-an-environment-module">Installing as an environment module</a></li>
</ul>
</li>
<li><a href="#running-multiqc">Running MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#choosing-where-to-scan">Choosing where to scan</a></li>
<li><a href="#renaming-reports">Renaming reports</a></li>
<li><a href="#overwriting-existing-reports">Overwriting existing reports</a></li>
<li><a href="#sample-names-prefixed-with-directories">Sample names prefixed with directories</a></li>
<li><a href="#using-different-templates">Using different templates</a></li>
<li><a href="#pdf-reports">PDF Reports</a></li>
<li><a href="#printing-to-stdout">Printing to stdout</a></li>
<li><a href="#parsed-data-directory">Parsed data directory</a></li>
<li><a href="#exporting-plots">Exporting Plots</a></li>
<li><a href="#choosing-which-modules-to-run">Choosing which modules to run</a></li>
</ul>
</li>
<li><a href="#using-multiqc-reports">Using MultiQC Reports</a>
<ul class="nav nav-stacked">
<li><a href="#browser-compatibility">Browser compatibility</a></li>
<li><a href="#report-layout">Report layout</a></li>
<li><a href="#general-statistics-table">General Statistics table</a></li>
<li><a href="#plots">Plots</a>
<ul class="nav nav-stacked">
<li><a href="#interactive-plots">Interactive plots</a></li>
<li><a href="#flat-plots">Flat plots</a></li>
<li><a href="#exporting-plots-1">Exporting plots</a></li>
<li><a href="#dynamic-plots">Dynamic plots</a></li>
</ul>
</li>
<li><a href="#toolbox">Toolbox</a>
<ul class="nav nav-stacked">
<li><a href="#highlight-samples">Highlight Samples</a></li>
<li><a href="#renaming-samples">Renaming Samples</a></li>
<li><a href="#hiding-samples">Hiding Samples</a></li>
<li><a href="#export">Export</a></li>
<li><a href="#save-settings">Save Settings</a></li>
</ul>
</li>
</li>
</ul>
</li>
<li><a href="#configuring-multiqc">Configuring MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#sample-name-cleaning">Sample name cleaning</a>
<ul class="nav nav-stacked">
<li><a href="#other-search-types">Other search types</a></li>
<li><a href="#clashing-sample-names">Clashing sample names</a></li>
</ul>
</li>
<li><a href="#module-search-patterns">Module search patterns</a></li>
<li><a href="#ignoring-files">Ignoring Files</a></li>
<li><a href="#large-sample-numbers">Large sample numbers</a>
<ul class="nav nav-stacked">
<li><a href="#disabling-on-load-plotting">Disabling on-load plotting</a></li>
<li><a href="#flat--interactive-plots">Flat / interactive plots</a></li>
<li><a href="#tables--beeswarm-plots">Tables / Beeswarm plots</a></li>
</ul>
</li>
</li>
</ul>
</li>
<li><a href="#customising-reports">Customising Reports</a>
<ul class="nav nav-stacked">
<li><a href="#titles-and-introductory-text">Titles and introductory text</a></li>
<li><a href="#report-logo">Report Logo</a></li>
<li><a href="#project-level-information">Project level information</a></li>
<li><a href="#order-of-modules">Order of modules</a></li>
<li><a href="#customising-tables">Customising tables</a></li>
<li><a href="#number-formatting">Number formatting</a></li>
<li><a href="#troubleshooting">Troubleshooting</a></li>
</ul>
</li>
<li><a href="#troubleshooting-1">Troubleshooting</a>
<ul class="nav nav-stacked">
<li><a href="#not-enough-samples-found">Not enough samples found</a></li>
<li><a href="#no-logs-found-for-a-tool">No logs found for a tool</a></li>
<li><a href="#error-messages-about-mkl-trial-mode--licences">Error messages about mkl trial mode / licences</a></li>
<li><a href="#locale-error-messages">Locale Error Messages</a></li>
</ul>
</li>
</li>
</ul>
</li>
<li><a href="#multiqc-modules">MultiQC Modules</a>
<ul class="nav nav-stacked">
<li><a href="#pre-alignment">Pre-alignment</a>
<ul class="nav nav-stacked">
<li><a href="#cutadapt">Cutadapt</a></li>
<li><a href="#fastqc">FastQC</a>
<ul class="nav nav-stacked">
<li><a href="#theoretical-gc-content">Theoretical GC Content</a></li>
</ul>
</li>
<li><a href="#fastq-screen">FastQ Screen</a></li>
<li><a href="#skewer">Skewer</a></li>
<li><a href="#trimmomatic">Trimmomatic</a></li>
</ul>
</li>
<li><a href="#aligners">Aligners</a>
<ul class="nav nav-stacked">
<li><a href="#bismark">Bismark</a></li>
<li><a href="#bowtie-1">Bowtie 1</a></li>
<li><a href="#bowtie-2">Bowtie 2</a></li>
<li><a href="#hicup">HiCUP</a></li>
<li><a href="#kallisto">Kallisto</a></li>
<li><a href="#star">STAR</a></li>
<li><a href="#salmon">Salmon</a></li>
<li><a href="#tophat">TopHat</a></li>
</ul>
</li>
<li><a href="#post-alignment">Post-alignment</a>
<ul class="nav nav-stacked">
<li><a href="#bamtools">Bamtools</a></li>
<li><a href="#bcftools">Bcftools</a></li>
<li><a href="#featurecounts">featureCounts</a></li>
<li><a href="#gatk">GATK</a>
<ul class="nav nav-stacked">
<li><a href="#varianteval">VariantEval</a></li>
</ul>
</li>
<li><a href="#htseq">HTSeq</a></li>
<li><a href="#methylqa">methylQA</a></li>
<li><a href="#peddy">Peddy</a></li>
<li><a href="#picard">Picard</a></li>
<li><a href="#preseq">Preseq</a></li>
<li><a href="#prokka">Prokka</a></li>
<li><a href="#qualimap">Qualimap</a></li>
<li><a href="#quast">QUAST</a></li>
<li><a href="#rseqc">RSeQC</a></li>
<li><a href="#samblaster">Samblaster</a></li>
<li><a href="#samtools">Samtools</a>
<ul class="nav nav-stacked">
<li><a href="#idxstats">idxstats</a></li>
</ul>
</li>
<li><a href="#slamdunk">Slamdunk</a></li>
<li><a href="#snpeff">SnpEff</a></li>
</ul>
</li>
</li>
</ul>
</li>
<li><a href="#custom-content">Custom Content</a>
<ul class="nav nav-stacked">
<li><a href="#introduction">Introduction</a></li>
<li><a href="#configuration">Configuration</a></li>
<li><a href="#data-formats">Data formats</a>
<ul class="nav nav-stacked">
<li><a href="#data-from-a-released-tool">Data from a released tool</a></li>
<li><a href="#data-as-part-of-multiqc-config">Data as part of MultiQC config</a></li>
<li><a href="#multiqc-specific-data-file">MultiQC-specific data file</a></li>
<li><a href="#separate-configuration-and-data-files">Separate configuration and data files</a></li>
</ul>
</li>
</li>
</ul>
</li>
<li><a href="#coding-with-multiqc">Coding with MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#writing-new-templates">Writing New Templates</a>
<ul class="nav nav-stacked">
<li><a href="#core-or-plugin">Core or plugin?</a></li>
<li><a href="#creating-a-template-skeleton">Creating a template skeleton</a></li>
<li><a href="#child-templates">Child templates</a></li>
<li><a href="#extra-init-variables">Extra init variables</a></li>
<li><a href="#jinja-template-variables">Jinja template variables</a></li>
<li><a href="#appendices">Appendices</a>
<ul class="nav nav-stacked">
<li><a href="#custom-plotting-functions">Custom plotting functions</a></li>
</ul>
</li>
</li>
</ul>
</li>
<li><a href="#writing-new-modules">Writing New Modules</a>
<ul class="nav nav-stacked">
<li><a href="#introduction-1">Introduction</a></li>
<li><a href="#core-modules--plugins">Core modules / plugins</a></li>
<li><a href="#initial-setup">Initial setup</a>
<ul class="nav nav-stacked">
<li><a href="#submodule">Submodule</a></li>
<li><a href="#entry-points">Entry points</a></li>
<li><a href="#multiqc-config">MultiQC config</a></li>
<li><a href="#documentation">Documentation</a></li>
<li><a href="#readme-and-changelog">Readme and Changelog</a></li>
<li><a href="#multiqcmodule-class">MultiqcModule Class</a></li>
<li><a href="#logging">Logging</a></li>
</ul>
</li>
<li><a href="#step-1-find-log-files">Step 1 - Find log files</a></li>
<li><a href="#step-2-parse-data-from-the-input-files">Step 2 - Parse data from the input files</a>
<ul class="nav nav-stacked">
<li><a href="#no-files-found">No files found</a></li>
<li><a href="#custom-sample-names">Custom sample names</a></li>
<li><a href="#identical-sample-names">Identical sample names</a></li>
<li><a href="#printing-to-the-sources-file">Printing to the sources file</a></li>
</ul>
</li>
<li><a href="#step-4-adding-to-the-general-statistics-table">Step 4 - Adding to the general statistics table</a></li>
<li><a href="#step-5-writing-data-to-a-file">Step 5 - Writing data to a file</a></li>
<li><a href="#step-6-create-report-sections">Step 6 - Create report sections</a></li>
<li><a href="#step-7-plot-some-data">Step 7 - Plot some data</a></li>
<li><a href="#appendices-1">Appendices</a>
<ul class="nav nav-stacked">
<li><a href="#adding-custom-css--javascript">Adding Custom CSS / Javascript</a></li>
</ul>
</li>
</li>
</ul>
</li>
<li><a href="#plotting-functions">Plotting Functions</a>
<ul class="nav nav-stacked">
<li><a href="#bar-graphs">Bar graphs</a>
<ul class="nav nav-stacked">
<li><a href="#switching-datasets">Switching datasets</a></li>
<li><a href="#interactive--flat-image-plots">Interactive / Flat image plots</a></li>
</ul>
</li>
<li><a href="#line-graphs">Line graphs</a>
<ul class="nav nav-stacked">
<li><a href="#switching-datasets-1">Switching datasets</a></li>
<li><a href="#additional-data-series">Additional data series</a></li>
</ul>
</li>
<li><a href="#scatter-plots">Scatter Plots</a></li>
<li><a href="#creating-a-table">Creating a table</a></li>
<li><a href="#beeswarm-plots-dot-plots">Beeswarm plots (dot plots)</a></li>
<li><a href="#heatmaps">Heatmaps</a></li>
<li><a href="#javascript-functions">Javascript Functions</a>
<ul class="nav nav-stacked">
<li><a href="#plotting-line-graphs">Plotting line graphs</a></li>
<li><a href="#plotting-bar-graphs">Plotting bar graphs</a></li>
<li><a href="#switching-counts-and-percentages">Switching counts and percentages</a></li>
<li><a href="#switching-plot-datasets">Switching plot datasets</a></li>
<li><a href="#custom-event-triggers">Custom event triggers</a></li>
</ul>
</li>
</li>
</ul>
</li>
<li><a href="#multiqc-plugins">MultiQC Plugins</a>
<ul class="nav nav-stacked">
<li><a href="#entry-points-1">Entry Points</a></li>
<li><a href="#modules">Modules</a></li>
<li><a href="#templates">Templates</a></li>
<li><a href="#command-line-options">Command line options</a></li>
<li><a href="#hooks">Hooks</a></li></ul></li></ul></li></ul>            <p class="backtotop"><a href="http://gitter.im/ewels/MultiQC">Discuss on Gitter</a></p>
            <p class="backtotop"><a href="#">Back to top</a></p>
          </div>
        </div>
      </div>
    </div> <!-- /container -->
    
    <footer id="footer">
      <div class="container">
        <p>Created by Phil Ewels:
          <a target="_blank" href="https://github.com/ewels"><i class="fa fa-github"></i> ewels</a> |
          <a target="_blank" href="https://twitter.com/tallphil"><i class="fa fa-twitter"></i> tallphil</a> |
          <a target="_blank" href="https://se.linkedin.com/in/philewels">LinkedIn</a> | 
          <a target="_blank" href="https://www.researchgate.net/profile/Philip_Ewels">ResearchGate</a>
        </p>
        <a target="_blank" href="http://www.scilifelab.se" class="scilife-footer">
          <img src="../images/SLL_logo_blue_white.png">
        </a>
      </div>
    </footer>


    <!-- Placed at the end of the document so the pages load faster -->
    <script src="../js/jquery-1.11.3.min.js"></script>
    <script src="../js/bootstrap.min.js"></script>
    <script src="../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="../js/docs.js"></script>
  
  <!-- Google Analytics -->
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-68098153-1', 'auto');
    ga('send', 'pageview');

  </script>
  
  </body>
</html>
